---
title: "Biostring"
author: "yincy"
date: "11/7/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load package, message=FALSE}
library(Biostrings)
```


## Overview  
The Biostrings package contains classes and functions for representing biological strings such as DNA, RNA and amino acids. In addition the package has functionality for pattern matching (short read alignment) as well as a pairwise alignment function implementing Smith-Waterman local alignments and Needleman-Wunsch global alignments used in classic sequence alignment (see (Durbin et al. 1998) for a description of these algorithms). There are also functions for reading and writing output such as FASTA files.  



## Representing Sequences  
There are two basic types of containers for representing strings. One container represents a single string (say a chromosome or a single short read) and the other container represnets a set of strings (say a set of short reads). There are different classes intended to represnet different types of sequences such as DNA or RNA sequences.  

```{r}
dna1 <- DNAString(x = "ACGT-N")
```

```{r}
DNAString(x = "ADE")
```

```{r}
dna2 <- DNAStringSet(x = c("ACGT", "GTCA", "GCTA"))
```

Note that the alphabet of a DNAString is an extended alphabet: - (for insertion) and N are allowed. In fact, IUPAC codes are allowed (these codes represent different characters, for example the code “M” represents either and “A” or a “C”).  

```{r IUPAC code}
IUPAC_CODE_MAP
```


Indexing into a DNAString retrieves a subsequence (similar to the standard R function substr), whereas indexing into a DNAStringSet gives you a subset of sequences.  

```{r}
dna1[2:4]
```

```{r}
dna2[2:3]
```

Note that [[ allows you to get a single element of a DNAStringSet as a DNAString. This is very similar to [ and [[ for lists.  

```{r}
dna2[[1]]
```


DNAStringSet objects can have names, like ordinary vectors  
```{r}
names(dna2) <- paste("seq", 1:3, sep = "")
dna2
```


The full set of string classes:  
- `DNAString[Set]`: DNA sequences.  
- `RNAString[Set]`: RNA sequences.  
- `AAString[Set]`: Amino acids sequences  
- `BString[Set]`: 'big' sequences, using any kind of letter.  

In addition you will often see references to XString[Set] in the documentation. An XString[Set] is basically any of the above classes.  

These classes seem very similar to standard characters() from base R, but there are important differences. The differences are mostly about efficiencies when you deal with either (a) many sequences or (b) very long strings (think whole chromosomes).  


## Basic functionality  
Basic character functionality is supported:  
- `length`, `names`.  
- `c` and `rev` (reverse the sequence).  
- `width`, `nchar` (number of characters in each sequence).  
- `==`, `duplicated`, `unique`.  
- `as.charcater` or `toString`: converts to a base character() vector.  
- `sort`, `order`.  
- `chartr`: convert some letters into other letters.  
- `subseq`, `subseq<-`, `extractAt`, `replaceAt`.  
- `replaceLetterAt`.  

```{r}
width(dna2)
```

9
```{r}
sort(dna2)
```

```{r}
rev(dna2)
```

## Biological functionality  
There are also functions which are related to the biological interpretation of the sequences, including  
- `reverse`: reverse the sequence  
- `complement`, `reverseComplement`: (reverse) complement the sequence  
- `translate`: translate the DNA or RNA sequence into amino acids  


```{r}
translate(dna2)
```

```{r}
dna1
reverseComplement(dna1)
complement(dna1)
```


## Counting letters  
We very often want to count sequences in various ways:  
- Compute the GC content of a set of sequences.  
- Compute the frequencies of dinucleotides in a set of sequences.  
- Compute a position weight matrix from a set of aligned sequences.  


There is a rich set of functions for doing this quickly.  
- `alphabetFrequency`, `letterFrequency`: Compute the frequency of all characters(`alphabetFrequency`) or only specific letters(`letterFrequency`).  

- `dinucleotideFrequency`, `trinucleotideFrequency`, `oligonucleotideFrequency`: compute frequencies of dinucleotides (2 bases), trinucleotides (3 bases) and oligonucleotides (general number of bases).  

- `letterFrequencyInSlidingView`: letter frequencies, but in sliding views along the string.  

- `consensusMatrix`: consensus matrix; almost a position weight matrix.  

```{r}
alphabetFrequency(dna1)
```


```{r}
alphabetFrequency(dna2)
```

```{r}
letterFrequency(x = dna2, letters = "GC")
```


```{r}
consensusMatrix(x = dna2, as.prob = TRUE)
```























