---
title: "rGADEM"
author: "yincy"
date: "3/31/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r, message=FALSE}
library(rGADEM)
library(BSgenome.Hsapiens.UCSC.hg38)
```


## Loading in the data  
The first step is to load data from BED files or FASTA format in the R environment. The sequences are stored in some of the basic containers defined in the Biostrings package. So, the data can be manipulated in a consistent and easy way.   

The maximum number of sequences allowed is 44000. But it is possible to change the default parameters by editing the defines.h file and recompiling it.  


### Load data from BED file  
each line on the BED file contain the location, start and end position on the chromosome.  

```{r}
pwd <- ""
path <- system.file("extdata/Test_100.bed", package = "rGADEM")
BedFile <- paste(pwd, path, sep = "")
BED <- read.table(BedFile, header = FALSE, sep = "\t")
BED <- data.frame(chr = as.factor(BED[, 1]), 
                  start = as.numeric(BED[, 2]), 
                  end = as.numeric(BED[, 3]))
```


```{r}
rgBED <- IRanges(
    start = BED[, 2], 
    end = BED[, 3]
)

Sequences <- GRanges(seqnames = BED[, 1], 
                     ranges = rgBED)
```

### Load from FASTA file  
```{r}
pwd <- ""
path <- system.file("extdata/Test_100.fasta", package = "rGADEM")
fastaFile <- paste(pwd, path, sep = "")
Sequences <- readDNAStringSet(fastaFile, format = "fasta")
```


## rGADEM analysis 
At this time, we are now ready to start rGADEM analysis.  

P-Value cutoff: The P-Value cutoff controls the number of binding site in a motif.  
E-Value cutoff: The E-Value cutoff controls the number of motifs to be identified.  
```{r}
gadem <- GADEM(Sequences = Sequences, 
               genome = Hsapiens, verbose = 1)
```


## Seeded analysis  
In a seeded analysis rGADEM does not generate the starting PWMs through spaced dyads and optimize them through a Genetic Algorithm. This makes seeded runs much faster than unseeded. The effciency of seeded runs makes it practical, even for sequence sets consisting of thousands of ChIP-seq peak cores, to assess several alternative seed PWMs, when prior knowledge suggests that this may be advisable. 

The main advantage of a seeded analysis over an unseeded analysis is its computational effciency.We recommend a seeded analysis whenever a reasonable starting PWM is available. However, for de novo motif discovery, an unseeded analysis is necessary.   

First step is to prepare a text fle with your PWM. It could be a general database (JASPAR, TransfacÂ©,...). Only STAT1 have been selected in our example but it is possible to select a list of PWMs.  

```{r}
path <- system.file("extdata/jaspar2009.txt", package = "rGADEM")
seededPWM <- readPWMfile(path)
STAT1.PWM <- seededPWM[grepl("STAT1", names(seededPWM))]
```
 
```{r}
gadem <- GADEM(Sequences = Sequences, 
               verbose = 1, 
               genome = Hsapiens, 
               Spwm = STAT1.PWM)
```


## rGADEM output  
gadem object is a list of object  

- align: this object contains the individual motifs identified and the location (seqID and position) of the sites in the original sequence data.  

- motif: This object contains PWM, motif consensus, motif length and aligned sequences for a specific motif.  

- parameters: this object contains parameters of rGADEM analysis.  


```{r}
slotNames(gadem)
```

```{r}
slot(gadem, "motifList") %>% .[[1]]
```

