---
title: "scran"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction  
Single-cell RNA sequencing (scRNA-seq) is a widely used technique for profiling gene expression in individual cells. This allows molecular biology to be studied at a resolution that cannot be matched by bulk sequencing of cell populations. The `scran` package implements methods to perform low-level processing of scRNA-seq data, including cell cycle phase assignment, scaling normalization, variance modelling and testing for corrrelated genes. This vignette provides brief descriptions of these methods and some toy examples to demonstrate their use.  

# Setting up the data  
We start off with a count matrix where each row is a gene and each column is a cell. These can be obtained by mapping read sequences to a reference genome, and then counting the number of reads mapped to the exons of each gene. (See, for example, the `Rsubread` package to do both of these tasks.) Alternatively, pseudo-alignment methods can be used to quantify the abundance of each transcript in each cell. For simplicity, we will pull out an existing dataset from the scRNAseq package.  

```{r}
library(scRNAseq)

sce.grun <- readRDS("../../Bioconductor/workflow/Single-Cell/scRNA-seq-online/data/sce.grun.rds")
```

This particular dataset is taken from a study of the human pancreas with the CEL-seq protocol (Grun et al. 2016). It is provided as a SingleCellExperiment object (from the SingleCellExperiment package), which contains the raw data and various annotations. We perform some cursory quality control to remove cells with low total counts or high spike-in percentages:  

```{r}
library(scuttle)

qcstats <- perCellQCMetrics(x = sce.grun)
qcfilter <- quickPerCellQC(x = qcstats, percent_subsets = "altexps_ERCC_percent")
sce.grun <- sce.grun[, !qcfilter$discard]

qcfilter %>% apply(2, summary)
```

# Normalizing cell-specific biases  
Cell-specific biases are normalized using the `computeSumFactors` method, which implements the deconvolution strategy for scaling normalization (A. T. Lun, Bach, and Marioni 2016). This computes size factors that are used to scale the counts in each cell. The assumption is that most genes are not differentially expressed (DE) between cells, such that any differences in expression across the majority of genes represents some technical bias that should be removed.  

```{r}
library(scran)
clusters <- quickCluster(sce.grun)
sce.grun <- computeSumFactors(x = sce.grun, clusters = clusters)

sizeFactors(sce.grun) %>% summary()
```

For larger data sets, clustering should be performed with the `quickCluster` function before normalization. Briefly, cells are grouped into clusters of similar expression; normalization is applied within each cluster to compute size factors for each cell; and the factors are rescaled by normalization between clusters. This reduces the risk of violating the above assumption when many genes are DE between clusters in a heterogeneous population. We also assume that quality control on the cells has already been performed, as low-quality cells with few expressed genes can often have negative size factor estimates.  

An alternative approach is to normalize based on the spike-in counts (Lun et al. 2017). The idea is that the same quantity of spike-in RNA was added to each cell prior to library preparation. Size factors are computed to scale the counts such that the total coverage of the spike-in transcripts is equal across cells. The main practical difference is that spike-in normalization preserves differences in total RNA content between cells, whereas `computeSumFactors` and other non-DE methods do not.  

```{r}
sce.grun2 <- computeSpikeFactors(x = sce.grun, spikes = "ERCC")
sizeFactors(sce.grun2) %>% summary
```

Regardless of which size factor calculation we choose, we will use the size factors to compute normalized expression values using logNormCounts() from scuttle. Each expression value can be interpreted as a log-transformed “normalized count”, and can be used in downstream applications like clustering or dimensionality reduction.  

```{r}
sce.grun <- logNormCounts(x = sce.grun)
```

# Variance modelling  
We identify genes that drive biological heterogeneity in the data set by modelling the per-gene variance. By only using a subset of highly variable genes in downstream analyses like clustering, we improve resolution of biological structure by removing uninteresting genes driven by technical noise. We decompose the total variance of each gene into its biological and technical components by fitting a trend to the endogenous variances (A. T. Lun, McCarthy, and Marioni 2016). The fitted value of the trend is used as an estimate of the technical component, and we subtract the fitted value from the total variance to obtain the biological component for each gene.  

```{r}
dec <- modelGeneVar(x = sce.grun)

plot(dec$mean, dec$total, xlab = "Mean log-expression", ylab = "Variance")
curve(expr = metadata(dec)$trend(x), col = "blue", add = T)
```

If we have spike-ins, we can use them to fit the trend instead. This provides a more direct estimate of the technical variance and avoids assuming that most genes do not exhibit biological variaility.  

```{r}
dec2 <- modelGeneVarWithSpikes(x = sce.grun, spikes = "ERCC")

plot(dec2$mean, dec2$total, xlab = "Mean log-expression", ylab = "Variance", pch = 19)
points(metadata(dec2)$mean, metadata(dec2)$var, col = "red", pch = 19) 
curve(metadata(dec2)$trend(x), col = "blue", add = T)
```

If we have some uninteresting factors of variation, we can block on these using  `block=`. This will perform the trend fitting and decomposition within each block before combining the statistics across blocks for output. Statistics for each individual block can also be extracted for further inspection.  

```{r}
dec3 <- modelGeneVar(x = sce.grun, block = sce.grun$donor)
per.block <- dec3$per.block

par(mfrow = c(3, 2))
for(i in seq_along(per.block)){
  decx <- per.block[[i]]
  plot(decx$mean, decx$total, xlab = "Mean log-expression", 
       ylab = "Variance", main = names(per.block)[[i]])
  curve(metadata(decx)$trend(x), col = "blue", add = T)
}
```

We can then extract some top genes for use in downstream procedures using the  `getTopHVGs()` function. A variety of different strategies can be used to define a subset of interesting genes:  

```{r}
# get the top 10% of genes
top.hvgs <- getTopHVGs(stats = dec, prop = 0.1)

# get the top 2000 genes
top.hvgs2 <- getTopHVGs(stats = dec, n = 2000)

# get all genes with positive biologcal components
top.hvgs3 <- getTopHVGs(stats = dec, var.threshold = 0)

# get all genes with FDR below 5%
top.hvgs4 <- getTopHVGs(stats = dec, fdr.threshold = 0.05)
```

The selected subset of genes can then be passed to the `subset.row` argument (or equivalent) in downstream steps. This process is demonstrated below for the PCA step, using the `runPCA()` function from the `scater` package.  

```{r}
library(scater)

sce.grun <- runPCA(x = sce.grun, subset_row = top.hvgs)
reducedDimNames(sce.grun)
```

# Automated PC choice  
Principal components analysis is commonly performed to denoise and compact the data prior to downstream analysis. A common question is how many PCs to retain; more PCs will capture more biological signal at the cost of retaining more noise and requiring more computational work. One approach to choosing the number of PCs is to use the technical component estimates to determine the proportion of variance that should be retained. This is implemented in `denoisePCA()`, which takes the estimates returned by `modelGeneVar()` or friends. (For greater accuracy, we use the fit with the spikes; we also subset to only the top HVGs to remove noise.)  

```{r}
sced <- denoisePCA(x = sce.grun, dec2, subset.row = getTopHVGs(dec2, prop = 0.1))
reducedDim(sced, "PCA") %>% ncol
```

Another approach is based on the assumption that each subpopulation should be separated from each other on a different axis of variation. Thus, we choose the number of PCs that is not less than the number of subpopulations (which are unknown, of course, so we use the number of clusters as a proxy). It is then a simple matter to subset the dimensionality reduction result to the desired number of PCs.  

```{r}
output <- getClusteredPCs(reducedDim(sce.grun))
npcs <- metadata(output)$chosen
reducedDim(sce.grun, "PCAsub") <- reducedDim(sce.grun, "PCA")[, 1:npcs, drop = FALSE]
```

# Graph-based clustering  















