---
title: "monocle3-Clustering-and-classifying-cells"
author: "yincy"
date: "4/8/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Clustering and classifying your cells  
Single-cell experiments are often performed on tissues containing many cell types. Monocle3 provides a simple set of functions you can use to group your cells according to their gene expression profiles into clusters. Often cells form clusters that correspond to one cell type or a set of highly related cell types.  

Monocle3 uses techniques to do this that are widely accepted in single-cell RNA-seq analysis and similar to the approaches used by `Seurat`, `scanpy` and other tools.   

In this section, you will learn how to cluster cells using Monocle 3. We will demonstrate the main functions used for clustering with the C. elegans data from Cao & Packer et al. This study described how to do single-cell RNA-seq with combinatorial indexing in a protocol called "sci-RNA-seq". Cao & Packer et al. used sci-RNA-seq to produce the first single-cell RNA-seq analysis of a whole animal, so there are many cell types represented in the data.  

```{r}
expression_matrix <- readRDS(file = "/home/yincy/git/Data/monocle/cao_l2_expression.rds")
cell_metadata <- readRDS(file = "/home/yincy/git/Data/monocle/cao_l2_colData.rds")
gene_metadata <- readRDS(file = "/home/yincy/git/Data/monocle/cao_l2_rowData.rds")

cds <- new_cell_data_set(
    expression_data = expression_matrix, 
    cell_metadata = cell_metadata, 
    gene_metadata = gene_metadata
)
```


### Pre-process the data  
Now that the data's all loaded up, we need to pre-process it. This step is where you tell Monocle 3 how you want to normalize the data, whether to use Principal Components Analysis (the standard for RNA-seq) or Latent Semantic Indexing (common in ATAC-seq), and how to remove any batch effects. We will just use the standard PCA method in this demonstration. When using PCA, you should specify the number of principal components you want Monocle to compute.  
```{r}
cds <- preprocess_cds(cds = cds, 
                      num_dim = 100, # the number of dimentions you want to return
                      method = "PCA", 
                      norm_method = "log") # normalization methods prior to reducing dimentionality
```

It's a good idea to check that you're using enough PCs to capture most of the variation in gene expression across all the cells in the data set. You can look at the fraction of variation explained by each PC using `plot_pc_variance_explained()`.  
```{r}
plot_pc_variance_explained(cds = cds)
```

We can see that using more than 100 pCs would capture only a small amount of additional variation, and each additional PC makes downstream steps in Monocle3 slower.  


### Reduce dimensionality and visualize the cells  
Now we're ready to visualize the cells. To do so, you can use either t-SNE, which is very popular in single-cell RNA-seq, or UMAP, which is increasingly common. Monocle 3 uses UMAP by default, as we feel that it is both faster and better suited for clustering and trajectory analysis in RNA-seq. To reduce the dimensionality of the data down into the X, Y plane so we can plot it easily, call `reduce_dimension()`  
```{r}
cds <- reduce_dimension(cds = cds, 
                        max_components = 2, 
                        reduction_method = "UMAP", 
                        preprocess_method = "PCA",
                        cores = 6)
```

As you can see the cells form many groups, some with thousands of cells, some with only a few. Cao & Packer annotated each cell according to type manually by looking at which genes it expresses. We can color the cells in the UMAP plot by the authors' original annotations using the `color_cells_by` argument to `plot_cells()`.  
```{r, message=FALSE, warning=FALSE}
plot_cells(cds = cds, 
           reduction_method = "UMAP",
           color_cells_by = "cao_cell_type", 
           group_cells_by = "partition",
           show_trajectory_graph = F, 
           norm_method = "size_only", 
           cell_size = 0.35, 
           rasterize = T)
```
 
Except for a few cases described in a moment, `color_cells_by` can be the name of any column in `colData(cds)`. Note that when `color_cells_by` is a categorical variable, labels are added to the plot, with each label positioned roughly in the middle of all the cells that have that label.  
```{r}
plot_cells(cds = cds,
           genes = c("cpna-2", "egl-21", "ram-2", "inos-1"), 
           show_trajectory_graph = F)
```

If you have a relatively large dataset (with >10,000 cells or more), you may want to take advantage of options that can accelerate UMAP. Passing `umap.fast_sgd=TRUE` to `reduce_dimension()` will use a fast stochastic gradient descent method inside of UMAP. If your computer has multiple cores, you can use the cores argument to make UMAP multithreaded. However, invoking `reduce_dimension()` with either of these options will make it produce slighly different output each time you run it. If this is acceptable to you, you could see signifant reductions in the running time of `reduction_dimension()`.  
```{r}
cds <- reduce_dimension(cds = cds, 
                        max_components = 2,
                        reduction_method = "tSNE", 
                        preprocess_method = "PCA")

plot_cells(cds = cds, 
           reduction_method = "tSNE", 
           color_cells_by = "cao_cell_type", 
           show_trajectory_graph = F)
```


### Check for and remove batch effects  
When performing gene expression analysis, it's important to check for batch effects, which are systematic differences in the transcriptome of cells measured in different experimental batches. These could be technical in nature, such as those introduced during the single-cell RNA-seq protocol, or biological, such as those that might arise from different litters of mice. How to recognize batch effects and account for them so that they don't confound your analysis can be a complex issue, but Monocle provides tools for dealing with them.  

you should always check for batch effects when you perform dimensionality reduction. You should add a column to the colData that encodes which batch each cell is from. Then you can simply color the cells by batch. Cao & Packer et al included a "plate" annotation in their data, which specifies which sci-RNA-seq plate each cell originated from. Coloring the UMAP by plate reveals  
```{r}
plot_cells(cds = cds, 
           color_cells_by = "plate", 
           label_cell_groups = FALSE)
```

Dramatic batch effects are not evident in this data. If the data contained more substantial variation due to plate, we'd expect to see groups of cells that really only come from one plate. Nevertheless, we can try and remove what batch effect is by running the align_cds() function:  
```{r}
cds <- align_cds(
    cds = cds, 
    num_dim = 100, 
    alignment_group = "plate", 
    preprocess_method = "PCA"
)

cds <- reduce_dimension(cds = cds, 
                        max_components = 2, 
                        reduction_method = "UMAP")

plot_cells(cds = cds, 
           color_cells_by = "plate", 
           label_cell_groups = F)
```

When run with the `alignment_group` argument, `align_cds()` tries to remove batch effects using mutual nearest neighbor alignment, a technique introduced by John Marioni's lab. Monocle 3 does so by calling Aaron Lun's excellent package `batchelor`.  


### Group cells into clusters  
Grouping cells into clusters is an important step in identifying the cell types represented in your data. Monocle uses a technique called `community detection` to group cells. This approach was introduced by Levine et al as part of the phenoGraph algorithm.   
```{r}
cds <- cluster_cells(
    cds = cds, 
    k = 10, 
    reduction_method = "UMAP"
)
```

Once you run `cluster_cells()`, the `plot_cells()` function will label each cluster of cells is labeled separately according to how you want to color the cells. For example, the call below colors the cells according to their cell type annotation, and each cluster is labeled according the most common annotation within it.  

```{r}
plot_cells(cds = cds, 
           color_cells_by = "partition", 
           label_cell_groups = F)
```

You can choose to label whole partitions instead of clusters by passing `group_cells_by="partition"`. You can also plot the top 2 labels per cluster by passing `labels_per_group=2` to `plot_cells()`. Finally, you can disable this labeling policy, making `plot_cells()` behave like it did before we called `cluster_cells()`.  

```{r}
plot_cells(
    cds = cds, 
    color_cells_by = "cao_cell_type", 
    label_groups_by_cluster = F
)
```


### FInd marker genes expressed by each cluster  
Once cells have been clustered, we can ask what genes makes them different from one another. To do that, start by calling the top_markers() function.  
```{r}
marker_test_res <- top_markers(cds = cds,
                               group_cells_by = "partition", 
                               reference_cells = 1000, 
                               cores = 6)
```

The data frame `marker_test_res` contains a number of metrics for how specifically expressed each gene is in each partition. We could group the cells according to cluster, partition, or any categorical variable in colData(cds). You can rank the table according to one or more of the specificity metrics and take the top gene for each cluster.  

```{r}
top_specific_markers <- marker_test_res %>% 
    filter(fraction_expressing >= 0.1) %>% 
    group_by(cell_group) %>% 
    top_n(n = 1, wt = pseudo_R2)

top_specific_markers %>% 
    pull(gene_short_name) %>% 
    duplicated() %>% table()

top_specific_marker_ids <- top_specific_markers %>% 
    pull(gene_id) %>% 
    unique()

plot_genes_by_group(cds = cds, 
                    markers = top_specific_marker_ids, 
                    group_cells_by = "partition", 
                    ordering_type = "maximal_on_diag", 
                    max.size = 3)
```

plot more marker genes  
```{r}
top_specific_markers <- marker_test_res %>% 
    filter(fraction_expressing >= 0.1) %>% 
    group_by(cell_group) %>% 
    top_n(n = 3, wt = pseudo_R2)

top_specific_markers %>% 
    pull(gene_short_name) %>% 
    duplicated() %>% table()

top_specific_marker_ids <- top_specific_markers %>% 
    pull(gene_id) %>% 
    unique()

plot_genes_by_group(cds = cds, 
                    markers = top_specific_marker_ids, 
                    group_cells_by = "partition", 
                    ordering_type = "cluster_row_col", 
                    max.size = 3)
```


### Annotate your cells according to type  
Identifying the type of each cell in your dataset is critical for many downstream analyses. There are several ways of doing this. One commonly used approach is to first cluster the cells and then assign a cell type to each cluster based on its gene expression profile. We've already seen how to use top_markers(). Reviewing literature associated with a marker gene often give strong indications of the identity of clusters that express it. In Cao & Packer >et al, the authors consulted literature and gene expression databases for markers restricted to each cluster in order to assign the identities contained in colData(cds)$cao_cell_type.  

To assign cell types based on clustering, we begain by creating a new column in `colData(cds)` and initialize it with the values of `clusters(cds`.  
```{r}
colData(cds)$assigned_cell_type <- as.character(partitions(cds))

colData(cds)$assigned_cell_type <- recode(
    .x = colData(cds)$assigned_cell_type, 
    "1" = "Germline", 
    "2" = "Body wall muscle", 
    "3" = "Unclassified neurons", 
    "4" = "Vulval precursors", 
    "5" = "Failed QC", 
    "6" = "Seam cells", 
    "7"="Pharyngeal epithelia",
    "8"="Coelomocytes",
    "9"="Am/PH sheath cells",
    "10"="Failed QC",
    "11"="Touch receptor neurons",
    "12"="Intestinal/rectal muscle",
    "13"="Pharyngeal neurons",
    "14"="NA",
    "15"="flp-1(+) interneurons",
    "16"="Canal associated neurons",
    "17"="Ciliated sensory neurons",
    "18"="Other interneurons",
    "19"="Pharyngeal gland",
    "20"="Failed QC",
    "21"="Ciliated sensory neurons",
    "22"="Oxygen sensory neurons",
    "23"="Ciliated sensory neurons",
    "24"="Ciliated sensory neurons",
    "25"="Ciliated sensory neurons",
    "26"="Ciliated sensory neurons",
    "27"="Oxygen sensory neurons",
    "28"="Ciliated sensory neurons",
    "29"="Unclassified neurons",
    "30"="Socket cells",
    "31"="Failed QC",
    "32"="Pharyngeal gland",
    "33"="Ciliated sensory neurons",
    "34"="Ciliated sensory neurons",
    "35"="Ciliated sensory neurons",
    "36"="Failed QC",
    "37"="Ciliated sensory neurons",
    "38"="Pharyngeal muscle"
)
```

```{r}
plot_cells(cds = cds, 
           group_cells_by = "partition", 
           color_cells_by = "assigned_cell_type")
```

Partition 7 has some substructure, and it's not obvious just from looking at the output of top_markers() what cell type or types it corresponds to. So we can isolate it with the choose_cells() function for further analysis.  

```{r}
cds_subset <- choose_cells(cds = cds)
```

Now we have a smaller cell_data_set object that contains just the cells from the partition we'd like to drill into. We can use graph_test() to identify genes that are differentially expressed in different subsets of cells from this partition.  
```{r}
pr_graph_test_res <- graph_test(cds = cds, 
                                neighbor_graph = "knn", 
                                cores = 3)

pr_deg_ids <- row.names(subset(pr_graph_test_res, morans_I > 0.01 & q_value < 0.05))
```


```{r}
gene_module_df <- find_gene_modules(
    cds = cds_subset[pr_deg_ids, ], 
    resolution = 1e-3
)
```

Plotting these modules' aggregate expression values reveala which cell express which modules 
```{r}
plot_cells(
    cds = cds_subset, 
    genes = gene_module_df, 
    label_cell_groups = F
)
```

You can explore the genes in each module or conduct gene ontology enrichment analysis on them to glean insights about which cell types are present. Suppose after doing this we have a good idea of what the cell types in the partition are. Let's recluster the cells at finer resolution and then see how they overlap with the clusters in the partition.  
```{r}
cds_subset <- cluster_cells(cds = cds_subset, 
                            resolution = 1e-2)

plot_cells(cds = cds_subset, 
           color_cells_by = "cluster")
```


Based on how the patterns line up, we'll make the following assignments  
```{r}
colData(cds_subset)$assigned_cell_type <- as.character(clusters(cds_subset)[colnames(cds_subset)])

colData(cds_subset)$assigned_cell_type <- dplyr::recode(colData(cds_subset)$assigned_cell_type,
                                                        "1"="Somatic gonad precursors",
                                                        "2"="Somatic gonad precursors",
                                                        "3"="Vulval precursors",
                                                        "4"="Sex myoblasts",
                                                        "5"="Sex myoblasts",
                                                        "6"="Vulval precursors",
                                                        "7"="Failed QC",
                                                        "8"="Vulval precursors",
                                                        "10"="Unclassified neurons",
                                                        "11"="Distal tip cells")

plot_cells(cds_subset, group_cells_by="cluster", color_cells_by="assigned_cell_type")
```

Now we can transfer the annotations from the cds_subset object back to the full dataset. We'll also filter out low-quality cells at this stage.  
```{r}
colData(cds)[colnames(cds_subset),]$assigned_cell_type <- colData(cds_subset)$assigned_cell_type

cds <- cds[,colData(cds)$assigned_cell_type != "Failed QC" | is.na(colData(cds)$assigned_cell_type )]

plot_cells(cds, group_cells_by="partition", 
           color_cells_by="assigned_cell_type", 
           labels_per_group=5)
```

### Automated annotation with Garnett  



