---
title: "ATACseqQC"
author: "yincy"
date: "3/29/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introductions  
Assay for Transposase-Accessible Chromatin using sequencing (ATAC-seq) is an alternative or complementary technique to MNase-seq, DNase-seq, and FAIRE-seq for chromatin accessibility analysis. The results obtained from ATAC-seq are similar to those from DNase-seq and FAIRE-seq. ATAC-seq is gaining popularity because it does not require cross-linking, has higher signal to noise ratio, requires a much smaller amount of biological material and is faster and easier to perform, compared to other techniques.  

To help researchers quickly assess the quality of ATAC-seq data, we have developed the ATACseqQC package for easily making diagnostic plots following the published guidelines. In addition, it has functions to preprocess ATACseq data for subsequent peak calling.   

# Quick start  
Here is an example using ATACseqQC with a subset of published ATAC-seq data. Currently, only bam input file format is supported.  

First install ATACseqQC and other packages required to run the examples. Please note that the example dataset used here is from human. To run analysis with dataset from a different species or differnt assembly, please install the corresponding BSgenome, TxDb and phastCons. For example, to analyze mouse data aligned to mm10, please install BSgenome.Mmusculus.UCSC.mm10, TxDb.Mmusculus.UCSC.mm10.knownGene and phastCons60way.UCSC.mm10. Please note that phstCons60way.UCSC.mm10 is optional, which can be obtained according to the vignettes of GenomicScores.  

```{r}
library(ATACseqQC)

bamfile <- system.file("extdata", "GL1.bam", package = "ATACseqQC", mustWork = TRUE)
bamfile.labels <- gsub(pattern = ".bam", replacement = "", x = basename(bamfile))
```


## Estimate the library complexity  
```{r, message=FALSE, warning=FALSE}
estimateLibComplexity(histFile = readsDupFreq(bamFile = bamfile))
```

## Fragment size distribution  
First, there should be a large proportion of reads with less than 100 bp, which represents the nucleosome-free region. Second, the fragment size distribution should have a clear periodicity, which is evident in the inset figure, indicative of nucleosome occupacy (present in integer multiples).  
```{r}
fragSize <- fragSizeDist(bamFiles = bamfile, bamFiles.labels = bamfile.labels)
```

## Nucleosome positioning  
### Adjust the read start sites  
Tn5 transposase has been shown to bind as a dimer and inserts two adaptors into accessible DNA locations separated by 9 bp.  

herefore, for downstream analysis, such as peak-calling and footprinting, all reads in input bamfile need to be shifted. The function shiftGAlignmentsList can be used to shift the reads. By default, all reads aligning to the positive strand are offset by +4bp, and all reads aligning to the negative strand are offset by -5bp.   

The adjusted reads will be written into a new bamfile for peak calling or footprinting.  

```{r}
possibleTag <- combn(x = LETTERS, m = 2)
```





















