---
title: "stimulated-v.s.-control-pmbc"
author: "yincy"
date: "1/20/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction  
The data is from this paper DOI: 10.1038/nbt.4042.     
In this experiment, PBMCs were split into a stimulated and control group and the stimulated group was treated with interferon beta. The response to interferon caused cell type specific gene expression changes that makes a joint analysis of all data difficult, with cells clustering both by stimulation condition and cell type. Here, we demonstrate our integration strategy, as described in [DOI: 10.1016/j.cell.2019.05.031], for performing integrated analyses to perform the identification of common cell types and enable comparative analysis.  

While this example demonstrates the integration of two datasets (conditions), these methods have been extended to multiple datasets. This [workflow](https://satijalab.org/seurat/pancreas_integration_label_transfer.html) provides an example of integrating four pancreatic islet datasets.  


## Integration goals  
The follwoing tutorial is designed to give you an overview of the kinds of comparative analysis on complex types that are possible using the Seurat integration procedure. Here is the three main goals:  

- Identify cell types that are present in both datasets.  
- Obtain cell type markers that are conserved in both control and stimulated cells.  
- Compare the datasets to find cell-type specific responses to stmulation.  


## Setup the Seurat object  
```{r, message=FALSE, warning=FALSE, error=FALSE}
library(Seurat)
library(cowplot)
```


```{r}

```


## Perform integration  
Identify anchors using the `FindIntegrationAnchors` function, which takes a list of Seurat objects as input, and use these anchors to integrate the two datasets together with `IntegrateData`.  
```{r}
immune.anchors <- FindIntegrationAnchors(object.list = ifnb.list, dims = 1:20)
immune.comnined <- IntegrateData(anchorset = immune.anchors, dims = 1:20)
```



## Perform an integrated analysis  
Now we can run a single integrated analysis on all cells!  

```{r}
DefaultAssay(immune.comnined) <- "integrated"

# Run the standard workflow for visualization and clustering  
immune.comnined <- ScaleData(object = immune.comnined, verbose = F)
immune.comnined <- RunPCA(object = immune.comnined, npcs = 30, verbose = F)

# t-SNE and Clustering  
immune.comnined <- RunUMAP(object = immune.comnined, reduction.key = "pca", dims = 1:20)
immune.comnined <- FindNeighbors(object = immune.comnined, reduction = "pca", dims = 1:20)
immune.comnined <- FindClusters(object = immune.comnined, resolution = 0.5)
```


Visualization  
```{r}
p1 <- DimPlot(object = immune.comnined, reduction = "umap", group.by = "stim")
p2 <- DimPlot(object = immune.comnined, reduction = "umap", label = T)
plot_grid(p1, p2)
```


To visualize the two conditions side-by-side, we can use the `split.by` argument to show each condition colored by cluster.  
```{r}
DimPlot(object = immune.comnined, reduction = "umap", split.by = "stim")
```



## Identify conserved cell type markers  
To identify canonical cell type marker genes that are conserved across conditions, we provide the `FindConservedMarkers` function. This function performs differential gene expression testing for each dataset/group and combines the p-values using meta-analysis methods from the MetaDE R package. For example, we can calculated the genes that are conserved markers irrespective of stimulation condition in cluster 7 (NK cells).  

```{r}
DefaultAssay(object = immune.comnined) <- "RNA"
nk.markers <- FindConservedMarkers(object = immune.comnined, ident.1 = 7, grouping.var = "stim", verbose = F)
heaad(nk.markers)
```


We can explore these marker genes for each cluster and use them to annotate our clusters as specific cell types.  
```{r}
FeaturePlot(object = immune.comnined, 
            features = c("CD3D", "SELL", "CREM", "CD8A", "GNLY", "CD79A", "FCGR3A", "CCL2", "PPBP"),
            min.cutoff = "q9")
```


```{r}
immune.comnined <- RenameIdents(object = immune.comnined, 
                                `0` = "CD14 Mono",
                                `1` = "CD14 Naive T",
                                `2` = "CD4 Memory T",
                                `3` = "CD16 Mono",
                                `4` = "B",
                                `5` = "CD8 T",
                                `6` = "T activated",
                                `7` = "NK",
                                `8` = "DC", 
                                `9` = "B Activated",
                                `10` = "MK",
                                `11` = "pDC",
                                `12` = "Eryth",
                                `13` = "Mono/MK Doublets")

DimPlot(object = immune.comnined, label = T)
```


The `DotPlot` function with the `split.by` parameter can be useful for viewing conserved cell type markers across conditions, showing both the expression level and the percentage of cells in a cluster expression any given gene.  
```{r}
Idents(object = immune.comnined) <- factor(Idents(object = immune.comnined), levels = c("Mono/Mk Doublets", "pDC",  "Eryth", "Mk", "DC", "CD14 Mono", "CD16 Mono", "B Activated", "B", "CD8 T", "NK", "T activated", "CD4 Naive T", "CD4 Memory T"))

marker.to.plot <- c("CD3D", "CREM", "HSPH1", "SELL", "GIMAP5", "CACYBP", "GNLY", "NKG7", "CCL5",  "CD8A", "MS4A1", "CD79A", "MIR155HG", "NME1", "FCGR3A", "VMO1", "CCL2", "S100A9", "HLA-DQA1", "GPR183", "PPBP", "GNG11", "HBA2", "HBB", "TSPAN13", "IL3RA", "IGJ")

DotPlot(object = immune.comnined, 
        features = rev(marker.to.plot), 
        cols = c("blue", "red"), 
        dot.scale = 8, 
        split.by = "stim") +
    RotatedAxis()
```



## Identify differential expressed genes across conditions  



    







