---
title: "AnnotationDbi"
author: "yincy"
date: "5/8/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# Introduction To Bioconductor Annotation Package  
```{r}
knitr::include_graphics(path = "figure/AnnotationDbi.PNG")
```

Gene centric AnnotationDbi packages include:  

+ Organism level: org.Mm.eg.db ...  
+ Platform level: hgu133plus2.db ...  
+ Homology leve: hom.Dm.inp.db ...  
+ System-biology level: Go.db ...  


Genome centric GenomicFeatures package include  

+ Transcriptome level: TxDb.Hsapiens.UCSC.hg19.knownGene ...  
+ Generic genome features: Can generate via GenomicFeatures  


One web-based resource access biomart, via the biomaRt package:  

+ Query web-based 'biomart' resource for genes, sequence, SNPs, and etc.  

The most popular annotation packages have been modified so that they can make use of new set of methods to more easily access their contents. These four methods are named: `columns`, `keytypes`, `keys`, and `select`.  


# AnnotationDb objects and the select method  
As previously mentioned, a new set of methods have been added that allow a simpler way of extracting identifier based annotations. All the annotation packages that support these new methods expose an object named exactly the same as the package itself. These objects are collectively called AnntoationDb objects for the class that they all inherit from. The more specific classes (the ones that you will actually see in the wild) have names like OrgDb, ChipDb or TxDb objects. These names correspond to the kind of package (and underlying schema) being represented. The methods that can be applied to all of these objects are `columns`, `keys`, `keytypes` and `select`.  

In addition, another accessor has recently been added which allows extraction of one column at a time. The `mapIds` method allows users to extract data into either a named character vector, a list or even a SimpleCharacterList. This method should work with all the different kinds of AnnotationDb object described below.  


# ChipDb objects and the select method  
```{r}
library(hgu95av2.db)

ls("package:hgu95av2.db")
```

```{r}
hgu95av2.db
```

If we want to know what kinds of data are retriveable via `select`, then we should use the `columns` method like this.  
```{r}
columns(hgu95av2.db)
```

If we further curious to know more about those values for columns, we can consult the help pages. 
```{r}
help("ACCNUM")
?"ACCNUM"
```


If we are curious about what kinds of fields we could potentiall use as keys to query the database, we can use the `keytypes` method. In a perfect world, this method will return values very similar to what was returned by `columns`, but in reality, some kinds of values make poor keys and so this list is often shorter.  
```{r}
keytypes(hgu95av2.db)
```

If we want to extract some sample keys of a particular type, we can use the `keys` method.  
```{r}
keys(hgu95av2.db, keytype = "SYMBOL", pattern = "^MT") %>% unique()
```


And finally, if we have some keys, we can use `select` to extract them. By simply using appropriate argument values with select we can specifiy what keys we want to look up values for (keys), what we want returned back (columns) and the type of keys that we are passing in (keytype).  
```{r}
# 1st get some example keys  
keytypes(hgu95av2.db)
k <- keys(hgu95av2.db, keytype = "PROBEID") %>% head()

# then call select 
select(x = hgu95av2.db, 
       keys = k, 
       columns = c("SYMBOL", "GENENAME"), 
       keytype = "PROBEID")
```

Finally if you wanted to extract only one column of data you could instead use the mapIds method like this.  
```{r}
k <- keys(x = hgu95av2.db, keytype = "PROBEID") %>% head()
mapIds(x = hgu95av2.db, 
       keys = k, 
       keytype = "PROBEID", 
       column = "GENENAME") %>% 
    as.data.frame()
```


# OrgDb objects and the select method  
An organism level package (an ‘org’ package) uses a central gene identifier (e.g.Entrez Gene id) and contains mappings between this identifier and other kinds of identifiers (e.g. GenBank or Uniprot accession number, RefSeq id, etc.). The name of an org package is always of the form org.<Ab>.<id>.db (e.g. org.Sc.sgd.db) where <Ab> is a 2-letter abbreviation of the organism (e.g. Sc for Saccharomyces cerevisiae) and <id> is an abbreviation (in lower-case) describing the type of central identifier (e.g. sgd for gene identifiers assigned by the Saccharomyces Genome Database, or eg for Entrez Gene ids).  

```{r}
library(org.Hs.eg.db)
org.Hs.eg.db
```


```{r}
columns(org.Hs.eg.db)
```

```{r}
keytypes(org.Hs.eg.db)
```

```{r}
k <- keys(x = org.Hs.eg.db, keytype = "UNIPROT")
select(x = org.Hs.eg.db, 
       keys = k, 
       keytype = "UNIPROT", 
       columns = c("SYMBOL", "PATH")) %>% 
    tibble::as_tibble()
```

```{r}
load(file = system.file("extdata", "resultTable.Rda", package = "AnnotationDbi"))

resultTable %>% head()
```

```{r}
annots <- select(x = org.Hs.eg.db, 
                 keys = rownames(resultTable), 
                 keytype = "ENTREZID", 
                 columns = c("SYMBOL", "GENENAME"))
merge(x = resultTable, y = annots, by.x = 0, by.y = "ENTREZID")
```

# Using select with GO.db  
When you load the GO.db package, a GODb object is also loaded. This allows you to use the columns, keys, keytypes and select methods on the contents of the GO ontology.  
```{r}
library(GO.db)

columns(GO.db)
GOIDs <- c("GO:0042254", "GO:0044183")
select(x = GO.db, 
       keys = GOIDs, 
       keytype = "GOID", 
       columns = "DEFINITION")
```


```{r}
keytypes(GO.db)
keys(GO.db, keytype = "TERM") %>% length()
```


# Using selct with TxDb packages  
A TxDb package (a ’TxDb’ package) connects a set of genomic coordinates to various transcript oriented features. The package can also contain Identifiers to features such as genes and transcripts, and the internal schema describes the relationships between these different elements. All TxDb containing packages follow a specific naming scheme that tells where the data came from as well as which build of the genome it comes from.  
```{r}
library(TxDb.Hsapiens.UCSC.hg38.knownGene)

TxDb.Hsapiens.UCSC.hg38.knownGene
```

```{r}
columns(TxDb.Hsapiens.UCSC.hg38.knownGene)
```

```{r}
keytypes(TxDb.Hsapiens.UCSC.hg38.knownGene)
```

```{r}
k <- keys(x = TxDb.Hsapiens.UCSC.hg38.knownGene, keytype = "TXID")

select(x = TxDb.Hsapiens.UCSC.hg38.knownGene, 
       keys = k, 
       keytype = "TXID", 
       columns = c("TXID", "TXSTART"))
```


# Using select with EnsDb packages  
Similar to the TxDb objects/packages discussed in the previous section, EnsDb objects/packages provide genomic coordinates of gene models along with additional annotations (e.g. gene names, biotypes etc) but are tailored to annotations provided by Ensembl. The central methods columns, keys, keytypes and select are all implemented for EnsDb objects. In addition, these methods allow also the use of the EnsDb specific filtering framework to retrieve only selected information from the database (see vignette of the ensembldb package for more information).  

```{r}
library(EnsDb.Hsapiens.v86)

EnsDb.Hsapiens.v86
```

```{r}
columns(EnsDb.Hsapiens.v86)
```


```{r}
keytypes(EnsDb.Hsapiens.v86)
```

```{r}
key <- keys(x = EnsDb.Hsapiens.v86, keytype = "GENEID")

select(EnsDb.Hsapiens.v86, 
       keys = key, 
       keytype = "GENEID", 
       columns = c("TXID", "TXSEQSTART", "TXBIOTYPE"))
```

We can modify the queries above to retrieve only genes encoded on chromosome Y. To this end we use filter objects available for EnsDb objects and its methods.  
```{r}
linkY <- keys(x = EnsDb.Hsapiens.v86, 
              filter = list(GeneBiotypeFilter(value = "lincRNA"), SeqNameFilter("Y")))

linkY %>% length()
```

```{r}
txs <- select(EnsDb.Hsapiens.v86, 
              keys = linkY, 
              keytype = "GENEID", 
              columns = c("TXID", "TXSEQSTART", "TXBIOTYPE"))
txs
```



