---
title: "SummarizedExperiment"
author: "yincy"
date: "11/11/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load packages, message=FALSE}
library(GenomicRanges)
library(airway)
```

The `SummarizedExperiment` class from `GenomicRanges` package, is an extention of the `ExpressionSet` class to include `GRanges`.  

## Start  
```{r}
data(airway)
airway
```

Some aspects of the object are very similar to ExpressionSet, although with slightly different names and types:  
- `colData` contains phenotype (sample) information, like `pData` for `ExpressionSet`. It returns a `DataFrame` instead of a `data.frame`  
```{r}
colData(airway)
```

use `$` to get a particular column  
```{r}
airway$cell
```


`colnames` are like `sampleNames` from `ExpressionSet`; `rownames` are like `featureNames`.  
```{r}
colnames(airway) %>% head(); rownames(airway) %>% head()
```


The measurement data are accessed by `assay` and `assays`. A `SummarizedExperiment` can contain multiple measurement matrices (all of the same dimension). You get all of them by `assays` and you seelct a particular one by `assays(object, name)`, where you can see the names when you print the object or by using `assayNames`.  
```{r}
assay(airway) %>% .[1:4, 1:4]
```


```{r}
assayNames(airway)
```

```{r}
assay(airway, "counts") %>% head()
```


So far, this is all information which could be stored in an ExpressionSet. The new thing is that SummarizedExperiment allows for a rowRanges (or granges) data representing the different features. The idea is that these GRanges tells us which part of the genome is summarized for a particular feature.  

```{r}
length(rowRanges(airway))
```

```{r}
dim(airway)
```


```{r}
rowRanges(airway)
```

```{r}
dim(airway)
```
`r nrow(airway)` genes, `r ncol(airway)` samples.  

See how rowRanges is a GRangesList (it could also be a single GRanges). Each element of the list represents a feature and the GRanges of the feature tells us the coordinates of the exons in the gene (or transcript). Because these are genes, for each GRanges, all the ranges should have the same strand and seqnames, but that is not enforced.  

```{r}
length(rowRanges(airway))
```

```{r}
lapply(airway, length) %>% sum()
```


For some operations, you don't need to use `rowRanges` first, you cna use the operation directly on the object.  
```{r}
start(rowRanges(airway))
```

```{r}
start(airway)
```

you can also use `grange` as synonymous for `rowRanges`.  

Subsetting works like `ExpressinSet`: there are two dimensions, the first diemention is features and the second dimension is samples.  

```{r}
airway[1:10, ]
```


Because the `SummarizedExperiment` contains a `GRanges[list]` you can also use `subsetByOverlaps`  
```{r}
gr <- GRanges(seqnames = "1",
              ranges = IRanges(start = 1, end = 10^7))

subsetByOverlaps(airway, gr)
```


##Summary  
phenotype data (sample information): `colData(airway)`  
access particular (column) phenotype data using `$`: `airway$cell`  
expression matrix: `assay(object, "name")`  
expression matrix name: `assayNames`  
sample names: `colnames`  
feature names: `rownames`  
Grange annotation data: `rowRanges` or `grange`  




