---
title: "GEOquery"
author: "yincy"
date: "11/11/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = F, warning = F)
```

```{r}
library(GEOquery)
library(magrittr)
```


NCBI Gene Expression Omnibus (GEO) has a lot of high-throughput genomics datasets publicly available. Despite the name, this database is not exclusively for gene expression data.  

## NCBI GEO  
NCBI GEO is organised as samples which are grouped into series. For bigger experiments there are both SubSeries and SuperSeries. A SuperSeries is all the experiments for a single paper; a SuperSeries can be decomposed into SubSeries which are different technologies. As an example, look at the SuperSeries GSE19486. In this paper they used 2 different platforms (this is a weird name; a platform is a combination of a technology and a species). And they did RNA-seq and ChIP-seq for two different factors (NFkB-II and Pol II). This results in 4 SubSeries (2 for RNA-seq and 2 for ChIP-seq).  


A simpler setup is for example GSE994 where samples from current and former smokers were compared, using an Affymetrix microarray.  

Data submitted to NCBI GEO can be both “raw” and “proceseed”. Let us focus on gene expression data for the moment. “Processed” data is normalized and quantified, typically at the gene level, and is usually provided in the form of a gene by sample matrix. “Raw” data can be anything, from sequencing reads to microarray image files. There can even be different states of “raw” data, for example for a RNA-seq dataset you may have  
- FASTQ files (raw reads)  
- BAM files (aligned reads)  
- gene by sample expression matrix (unnormalized)  
- gene by sample expression matrix (normalized)  

In some cases there is a consensus in the field. For Affymetrix gene expression microarrays, “raw” files are so-called CEL files (a file format invented by Affymetrix) and “processed” data is normalized and quantified data, summarized at the probeset level.  

At the end of the day, GEO has series identifiers (like GSE19486) and sample identiers (GSM486297). Note the GSE vs GSM in the same. A user is almost always interested in all the samples in a given series, so the starting point is the series identifier, also called the accession number.  



## GEOquery  
```{r, message=FALSE, warning=FALSE}
eList <- getGEO(GEO = "GSE11675")
```

```{r}
class(eList)
```

```{r}
length(eList)
```

```{r}
names(eList)
```

```{r}
eData <- eList[[1]]
```

The function returns a list because you might be getting multiple SubSeries. In this case there is only one, and the list element is an ExpressionSet ready for usage! The phenotype data (which GEO knows about) is contained inside the pData slot.  

```{r}
colnames(pData(eData))
```

```{r}
dim(exprs(eData))
exprs(eData) %>% head()
```


However, what we got here was processed data. Users (including me) often wants access to more raw data. This is called “supplementary files” in GEO language and we can get those as well.  

```{r}
eList <- getGEOSuppFiles(GEO = "GSE11675")
```

## Other packages  
- `SRAdb` for Short Read Archive (SRA)  
- `ArrayExpress` a similar database to NCBI GEO but hosted at the EBI  










