---
title: "BSgenome"
author: "yincy"
date: "11/7/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r}
library(BSgenome)
library(BSgenome.Scerevisiae.UCSC.sacCer2)
library(magrittr)
```


## Overview  
The `BSgenome` package contains infrastructure for representing genome sequences in Bioconductor.  


## Genomes  
The BSgenome package provides support for genomes. In Bioconductor, we have special classes for genomes, because the chromosomes can get really big. For example, the human genome takes up several GB of memory.  

The available.genomes() function lists which genomes are currently available from from Bioconductor (it is possible to make your own genome package). Note that there are several so-called “masked” genomes, where some parts of the genome are masked. We will avoid this subject for now.  

```{r}
available.genomes() %>% head()
```


```{r}
library(BSgenome.Scerevisiae.UCSC.sacCer2)
Scerevisiae
```

A `BSgenome` package contains a single object which is the second component of the name. At first, nothing is loaded into memory, which makes it very fast. You can get the length and names of the chromosomes without actually loading them.  

```{r}
seqlengths(Scerevisiae)
```


```{r}
seqnames(Scerevisiae)
```

We load a chromosome by using the [[ or $ operators.  
```{r}
Scerevisiae$chrI
```


We can now do things like compute the GC content of the first chromosome  
```{r}
letterFrequency(Scerevisiae$chrI, "CG", as.prob = T)
```


To iterate over chromosomes seems straightforward with `lapply`. However, this function may end up using a lot of memory because the entire genome is loaded. Instead there is the `bsapply` function which handles loading and unloading of different chromosomes. The interface to `bsapply` is weird at first; you set up a `BSparams` object which contains which function you are using and which genome you are using it on (and a bit more information).  

```{r}
param <- new(Class = "BSParams", X = Scerevisiae, FUN = letterFrequency)
bsapply(param, letters = "GC") %>% head()
```


Note how the additional argument `letters` to the `letterFrequency` function is given as an argument to `bsapply`, not to the `BSParams` object.  

you can simplify the output (like `lapply` and `sapply`)  
```{r}
param <- new(Class = "BSParams", X = Scerevisiae, FUN = letterFrequency, simplify = T)
bsapply(param, letters = "CG")
```


```{r}
sum(bsapply(param, letters = "GC")) / sum(seqlengths(Scerevisiae))
```


# Views  
```{r}
library(BSgenome)
library(BSgenome.Scerevisiae.UCSC.sacCer2)
library(AnnotationHub)
```

Views are used when you have a single big object (think chromosome or other massive dataset) and you need to deal with (many) subsets of this object. Views are not restricted to genome sequences.  

Technically, a Views is like an IRanges couple with a pointer to the massive object. The IRanges contains the indexes.  

```{r}
dnaseq <- DNAString(x = "ACGTACGT")
vi <- matchPattern(pattern = dnaseq, subject = Scerevisiae$chrI)
```

get the IRange component  
```{r}
ranges(vi)
```

```{r}
Scerevisiae$chrI[start(vi):end(vi)]
```


the `Views` object also look a bit like a `DNAStringSet`  
```{r}
alphabetFrequency(vi)
```

The advantage of Views is that they don’t duplicate the sequence information from the subject; all they keep track of are indexes into the subject (stored as IRanges). This makes it very (1) fast, (2) low-memory and makes it possible to do things like  
```{r}
shift(x = vi, shift = 10)
```

where we now get the sequence 10 bases next to the original match. This could not be done if all we had were the bases of the original subsequence.  


Views are especially powerful when there are many of them. A usecase I often have are the set of all exons (or promoters) of all genes in the genome. You can use GRanges as Views as well.  
```{r}
gr <- vmatchPattern(pattern = dnaseq, subject = Scerevisiae)
vi2 <- Views(Scerevisiae, gr)
```

```{r}
ahub <- AnnotationHub()
qh <- query(ahub, c("sacCer2", "genes"))
```

```{r}
genes <- qh[[which(qh$title == "SGD Genes")]]
```

compute the GC content of all promoters in the yeast genome  
```{r}
prom <- promoters(genes)
head(prom, n = 3)
```

We get a warning that some of these propoters are out-of-band.  
```{r}
prom <- trim(prom)
promViews <- Views(subject = Scerevisiae, prom)
gcProm <- letterFrequency(x = promViews, letters = "GC", as.prob = TRUE)

head(gcProm)
```


```{r}
params <- new(Class = "BSParams", X = Scerevisiae, FUN = letterFrequency, simplify = TRUE)
gccontent <- bsapply(params, letters = "GC")

gcPercentage <- sum(gccontent) / sum(seqlengths(Scerevisiae))
gcPercentage
```


compare this genome percentage to the distribution of GC content for promoters  
```{r}
plot(density(gcProm))
abline(v = gcPercentage, col = "red")
```


















