---
title: "ExpressionSet"
author: "yincy"
date: "11/11/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load packages}
library(Biobase)
library(ALL)
library(hgu95av2.db)
library(magrittr)
```


## Overview  
`ExpressionSet` class is a fundamental data container in Bioconductor, the class is define in the `Biobase` package.  

This class has inspired many other data containers in Bioconductor and can be considered a great success in programming with data.  


## start  
```{r}
data(ALL)
```

This is an experiment done on an Affymetrix HGU 95av2 gene expression microarray; the authors profiled 128 samples.  

```{r}
experimentData(ALL)
```


The core of the object are two matrices   
- the `exprs` matrix containing the 12625 gene expression measurements on the 128 samples  
- the `pData data.frame` containing phenotype data on the samples.  

```{r}
exprs(ALL) %>% .[1:4, 1:4]
```


```{r}
sampleNames(ALL) %>% head()
```


```{r}
featureNames(ALL) %>% head()
```

```{r}
pData(ALL)
```


access individual columns of this `data.frame` by using the `$` operator.  
```{r}
pData(ALL)$age %>% head()
```

```{r}
ALL$age %>% head()
```


## Subsetting  
Subsetting of this object is an important operation. The subsetting has two dimensions; the first dimension is genes and the second is samples. It keeps track of how the expression measures are matched to the pheno data.  

```{r}
ALL[, 1:5]
print("-----------")
ALL[1:5, ]
```

## featureData and annotation  
You can think of `pData` as providing information on the columns (samples) of the measurement matrix. Similar to `pData`, we have `featureData` which is meant to provide information on the features (genes) of the array.  

```{r}
featureData(ALL)
```

```{r}
ids <- featureNames(ALL)
```

```{r}
as.list(hgu95av2ENTREZID[ids]) %>% head()
```


This gives us the Entrez ID associated with the different measurements. There are a number of so-called “maps” like hgu95av2XX with XX having multiple values. This approach is very specific to Affymetrix chips. An alternative to using annotation packages is to use the biomaRt package to get the microarray annotation from Ensembl (an online database).  


## Note: phenoData and pData  
For this type of object, there is a difference between `phenoData` (an objects of class `AnnotatedDataFrame`) and `pData` (an object of class `data.frame`).  

The idea behind `AnnotatedDataFrame` was to include additional information on what a `data.frame` contains, by having a list of descriptions called `varLabels`.   

```{r}
pD <- phenoData(ALL)
varLabels(pD)
```

But these days, it seems that varLabels are constrained to be equal to the column names of the data.frame making the entire AnnotatedDataFrame construction unnecessary  


## The eSet class  
The ExpressionSet class is an example of an eSet. The ExpressionSet class has a single measurement matrix which we access by exprs. In contrast to this, the eSet class can have any number of measurement matrices with arbitrary names, although all matrices needs to have the same dimension.  

An example of another eSet derived class is NChannelSet which is meant to store multichannel microarray measurements; an example is a two-color array where the microarray has been imaged in two different colors.  

Another example of classes are from the minfi package for DNA methylation arrays; here there are classes such as RGChannelSet for a two color microarray and MethylSet for methylation measurements.  

Another example of classes are from the minfi package for DNA methylation arrays; here there are classes such as RGChannelSet for a two color microarray and MethylSet for methylation measurements.  







































