---
title: "Analysis-of-ATAC-seq-Data"
author: "yincy"
date: "12/12/2019"
output: 
  html_document:
    toc: true
    toc_depth: 2
    number_section: true
    highlight: tango
    code_folding: show
    smooth_scroll: true
    
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = F, warning = F, error = F)
```

```{r, fig.align='center', fig.cap="ATAC-seq, MNase-seq and DNase-seq"}
knitr::include_graphics(path = "figure/chromatin-accessibility assays.PNG")
```

```
DNase-seq - Enzymatic digestion to extract signal from open chromatin around transcription factor binding sites.  

MNase-seq - Enzymatic digestion to extract signal repesenting nucleosome positioning.  

ATAC-seq - Uses transposases and offers a method to simultaneously extract signal from transcription factors binding sites and nucleosome positions from a single sample.  
```

# Working with ATAC-seq data in R/Bioconductor  
## Introdcution  
data sets:
- GSM1155958  
- Liver day 12  
- Kidney day 15  
- Hindbrain day 12  
- T-reg  

required packages  
```
library(knitr)
library(rmdformats)
library(DT)
library(tidyverse)
library(magrittr)
library(devtools)
library(Rsubread)
library(Rsamtools)
library(GenomicAlignments)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(soGGi)
library(rtracklayer)
library(ChIPQC)
library(ChIPseeker)
library(rGREAT)
library(limma)
library(DESeq2)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(tracktables)
library(MotifDb)
library(Biostrings)
library(BSgenome)
```

## Process data  
### Buildindex  
Using the already build index from NCBI mouse GRCm38  

### Aligning Sequence Reads to the genome  
```{r}
read1 <- "/home/yincy/Documents/ATAC/GSM1155958/fastq/SRR891269_1.fastq.gz"
read2 <- "/home/yincy/Documents/ATAC/GSM1155958/fastq/SRR891269_2.fastq.gz"
outBAM <- "/home/yincy/Documents/ATAC/GSM1155958/BAM/ATAC_50k_2.bam"

align(index = "/home/yincy/Documents/genome/subread_index/ENCODE/human/GRCh38.encode",
      readfile1 = read1,
      readfile2 = read2,
      output_format = "BAM",
      output_file = outBAM,
      nthreads = 14,
      type = 1,
      unique = TRUE,
      maxFragLength = 1000)
```


### Sorting and Indexing  
```{r}
library(Rsamtools)
sortedBAM <- file.path(dirname(path = outBAM), paste0("Sorted_", basename(outBAM)))

sortBam(file = outBAM, 
        destination = gsub("\\.bam",  "", basename(sortedBAM)))

indexBam(files = sortedBAM)
```

```{r}
sortedBAM <- "/home/yincy/Documents/ATAC/GSM1155958/BAM/Sorted_ATAC_50k_2.bam"
```

### Number of mapped reads  
```{r}
library(Rsubread)
pmapped <- propmapped(files = sortedBAM)
```


### Distribution of mapped reads  
check the distribution of ampped reads across chromosomes.  
ATAC-seq is known have high signal on the mitochondrial chromosomes and check out for this.  

```{r}
library(Rsamtools)
library(ggplot2)
library(magrittr)
library(dplyr)
```

```{r}
idxstatsBam(file = sortedBAM) %>% 
  arrange() %>% 
  head(n = 25) %>% 
  ggplot(aes(seqnames, mapped)) +
  geom_bar(aes(fill = seqnames), stat = "identity", show.legend = F) +
  scale_y_continuous(name = "mapped reads") +
  scale_x_discrete(label = c(paste("chr", 1:22, sep = ""), "chrX", "chrY", "chrM")) +
  coord_flip()
```


## Dataset post-alignment processing  
### reading in mapped reads  
```{r}
library(GenomicAlignments)

atacReads <- readGAlignmentPairs(file = sortedBAM,
                                 param = ScanBamParam(mapqFilter = 1, flag = scanBamFlag(isPaired = T, isProperPair = T),
                                                      what = c("qname", "mapq", "isize"),
                                                      which = GRanges(seqnames =  "NC_000020.11", 
                                                                      ranges = IRanges(start = 1, end = 64444167))))
```


### Retrieving insert sizes  
```{r}
atacReads_read1 <- GenomicAlignments::first(atacReads)
insertSizes <- abs(mcols(atacReads_read1)$isize)
head(insertSizes)
```


### Plotting insert sizes  
```{r}
library(magrittr)
library(dplyr)
library(ggplot2)

table(insertSizes) %>% 
  data.frame() %>% 
  rename(InsertSize = insertSizes, Count = Freq) %>% 
  mutate(InsertSize = as.numeric(as.vector(InsertSize)),
         Count = as.numeric(as.vector(Count))) %>% 
  ggplot(aes(InsertSize, Count)) +
  geom_line() +
  theme_classic() +
  scale_y_continuous(trans = "log2") +
  geom_vline(xintercept = c(180, 247), color = "red") +
  geom_vline(xintercept = c(315, 437), color = "blue") +
  geom_vline(xintercept = 100, color = "darkgreen")
```


### Plotting ATAC-seq signal of TSSs (Retrieving TSSs regions)  
From the fragment length plots shown above it would seem we have evidence of data coming both from nucleosome occupied and nucleosome free regions.  

We expect nucleosome free regions to be present at TSSs of active genes and nucleosome signal to be more enriched in regions surrounding the TSS.  

First then we need a set of TSS positions which we can retrieve from the gene modles in the **TxDb.Hsapiens.UCSC.hg38.knownGene** package.  
```{r}
library("TxDb.Hsapiens.UCSC.hg38.knownGene")
TSSs <- resize(x = genes(TxDb.Hsapiens.UCSC.hg38.knownGene), fix = "start", width = 1)
```


### Plotting ATAC-seq signal of TSSs 
>Creating open, mono- and dinucleosome signal profiles  

```{r}
library(soGGi)

# Nucleosome free  
nucFree <- regionPlot(bamFile = sortedBAM, 
                      testRanges = TSSs, 
                      style = "point", 
                      format = "bam", 
                      paired = TRUE,
                      minFragmentLength = 0, 
                      maxFragmentLength = 100,
                      forceFragment = 50)

# Mononucleosome  
monoNuc <- regionPlot(bamFile = sortedBAM,
                      testRanges = TSSs,
                      style = "point",
                      format = "bam",
                      paired = TRUE,
                      minFragmentLength = 180,
                      maxFragmentLength = 240,
                      forceFragment = 80)

# dinucleosome  
diNuc <- regionPlot(bamFile = sortedBAM, 
                    testRanges = TSSs,
                    style = "point",
                    format = "bam",
                    paired = TRUE,
                    minFragmentLength = 315, 
                    maxFragmentLength = 437,
                    forceFragment = 160)
```



















