---
title: "nichenetr"
author: "yincy"
date: "4/21/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# nichenetr  
**nichenetr**: **the implementation of the NicheNet method**. The goal of NicheNet is to study intercellular communication from a computational perspective. NicheNet uses human or mouse gene expression data of interacting cells as input and combines this with a prior model that integrates existing knowledge on ligand-to-target signaling paths. This allows to predict ligand-receptor interactions that might drive gene expression changes in cells of interest.  

# Introduction to NicheNet  
The figure below shows a graphical representation of the NicheNet workflow. Interactions inferred from several complementary ligand-receptor, signaling and gene regulatory data sources were aggregated in respective integrated networks from which ligand-target regulatory potential scores were calculated. This model of prior information on potential ligand-target links can then be used to infer active ligand-target links between interacting cells. NicheNet prioritizes ligands according to their activity (i.e., how well they predict observed changes in gene expression in the receiver cell) and looks for affected targets with high potential to be regulated by these prioritized ligands.  
```{r NicheNet workflow}
knitr::include_graphics("/home/yincy/git/nichenetr/vignettes/workflow_nichenet.png")
```

NicheNet strongly differs from most current computational approaches to study intercellular communication. Current approaches study intercellular communication from (single-cell) expression data by linking ligands expressed by sender cells to their corresponding receptors expressed by receiver cells. **However, functional understanding of a cellular communication process also requires knowing how these inferred ligand-receptor interactions result in changes in the expression of downstream target genes within the receiver cells**.  

To address this need, we developed NicheNet. Contrary to existing approaches, NicheNet looks at gene regulatory effects of ligands because the used prior knowledge goes beyond ligand-receptor interactions and incorporates intracellular signaling and transcriptional regulation as well. As a result, NicheNet allows to predict which ligands influence the expression in another cell, which target genes are affected by each ligand and which signaling mediators may be involved. By generating these novel types of hypotheses, NicheNet can drive an improved functional understanding of a cell-cell communication process of interest. The figure below summarizes the conceptual differences between most current ligand-receptor network inference approaches (top panel) and NicheNet (bottom panel) and visualizes the power of NicheNet in prioritizing ligand-receptor interactions based on gene expression effects.  
```{r Comparing with other method}
knitr::include_graphics("/home/yincy/git/nichenetr/vignettes/comparison_other_approaches_2.png")
```

# Main functionalities of nichenetr  
Specific functionalities of this package include  

- assessing how well ligands expressed by a sender cell can predict changes in gene expression in the receiver cell.  
- prioritizing ligands based on their effect on gene expression.  
- inferring putative ligand-target links active in the system under study.  
- inferring potential signaling paths between ligands and target genes of interest: to generate causal hypotheses and check which data sources support the predictions.  
- validation of the prior ligand-target model.  
- construction of user-defined prior ligand-target models.  


# Main Steps  
1. define expressed genes in sender and receiver cell populations  
2. define gene set of interest, the genes in reciver might be affected by sender cell expressed ligands.  
3. define the potential ligands that might affect the gene set of interest `predict_ligand_activities()`.  
4. infer targeted genes of top-ranked ligands `get_weighted_ligand_target_links()`.  
5. inferring ligand-to-target signaling paths `get_ligand_signaling_path()`.  
6. infer supporting datasources `infer_supporting_datasources()`.  



