---
title: "021_Perform_NicheNet-analysis-starting-from-a-Seurat-object:step-by-step"
author: "yincy"
date: "4/15/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction  
In this vignette, you can learn how to perform a basic NicheNet analysis on a Seurat v3 object. Such a NicheNet analysis can help you to generate hypotheses about an intercellular communication process of interest for which you have single-cell gene expression data as a Seurat object. Specifically, NicheNet can predict 1) which ligands from one or more cell population(s) (“sender/niche”) are most likely to affect target gene expression in an interacting cell population (“receiver/target”) and 2) which specific target genes are affected by which of these predicted ligands.  

Because NicheNet studies how ligands affect gene expression in putatively neighboring/interacting cells, you need to have data about this effect in gene expression you want to study. So, there need to be ‘some kind of’ differential expression in a receiver cell population, caused by ligands from one of more interacting sender cell populations.  

In this vignette, we demonstrate the use of NicheNet on a Seurat Object. The steps of the analysis we show here are also discussed in detail in the main, basis, NicheNet vignette NicheNet’s ligand activity analysis on a gene set of interest: predict active ligands and their target genes:vignette("ligand_activity_geneset", package="nichenetr"). Make sure you understand the different steps in a NicheNet analysis that are described in that vignette before proceeding with this vignette and performing a real NicheNet analysis on your data. This vignette describes the different steps behind the wrapper functions that are shown in Perform NicheNet analysis starting from a Seurat object:vignette("seurat_wrapper", package="nichenetr"). Following this vignette has the advantage that it allows users to adapt specific steps of the pipeline to make them more appropriate for their data.  

As example expression data of interacting cells, we will use mouse NICHE-seq data from Medaglia et al. to explore intercellular communication in the T cell area in the inguinal lymph node before and 72 hours after lymphocytic choriomeningitis virus (LCMV) infection (See Medaglia et al. 2017). We will NicheNet to explore immune cell crosstalk in response to this LCMV infection.  

In this dataset, differential expression is observed between CD8 T cells in steady-state and CD8 T cells after LCMV infection. NicheNet can be applied to look at how several immune cell populations in the lymph node (i.e., monocytes, dendritic cells, NK cells, B cells, CD4 T cells) can regulate and induce these observed gene expression changes. NicheNet will specifically prioritize ligands from these immune cells and their target genes that change in expression upon LCMV infection.  

# Prepare NicheNet analysis  
## loading required packages and data  
The NicheNet ligand-receptor network and weighted networks are necessary to define and show possible ligand-receptor interactions between two cell populations. The ligand-target matrix denotes the prior potential that particular ligands might regulate the expression of particular target genes. This matrix is necessary to prioritize possible ligand-receptor interactions based on observed gene expression effects (i.e. NicheNet’s ligand activity analysis) and infer affected target genes of these prioritized ligands.  

```{r, message=FALSE, warning=FALSE}
library(nichenetr)
library(Seurat)
library(tidyverse)
```

## Read in the expression data of interacting cells  
The dataset used here is publicly available single-cell data from immune cells in the T cell area of the inguinal lymph node. The data was processed and aggregated by applying the Seurat alignment pipeline. The Seurat object contains this aggregated data. Note that this should be a Seurat v3 object and that gene should be named by their official mouse/human gene symbol.  

```{r}
seuratObj <- readRDS("../../Data/NicheNet/seuratObj.rds")
seuratObj[[]] %>% head()
```

Visualize which cell populations are present: CD4 T cells (including regulatory T cells), CD8 T cells, B cells, NK cells, dendritic cells (DCs) and inflammatory monocytes  

```{r}
seuratObj[[]] %>% pull(celltype) %>% table

DimPlot(seuratObj, reduction = "tsne")
```

Visualize the data to see to which condition cells belong. The metadata dataframe column that denotes the condition (steady-state or after LCMV infection) is here called ‘aggregate.’  

```{r}
seuratObj[[]] %>% pull(aggregate) %>% table

DimPlot(seuratObj, reduction = "tsne", group.by = "aggregate")
```

## read in Nichenet's model  
```{r}
ligand_target_matrix <- readRDS("../../Data/NicheNet/ligand_target_matrix.rds")
ligand_target_matrix[1:5, 1:5]
```

```{r}
lr_network <- readRDS("../../Data/NicheNet/lr_network.rds")
lr_network %>% head
```

```{r}
weighted_network <- readRDS("../../Data/NicheNet/weighted_networks.rds")
weighted_network_lr <- weighted_network$lr_sig %>% 
    inner_join(lr_network %>% distinct(from, to), by = c("from", "to"))

weighted_network$lr_sig %>% head
weighted_network$gr %>% head
```

Because the expression data is of mouse origin, we will convert the NicheNet network gene symbols from human to mouse based on one-to-one orthology:  

```{r}
lr_network <- lr_network %>% mutate(from = convert_human_to_mouse_symbols(from), 
                                    to = convert_human_to_mouse_symbols(to))

colnames(ligand_target_matrix) <- ligand_target_matrix %>% 
  colnames() %>% 
  convert_human_to_mouse_symbols()
rownames(ligand_target_matrix) <- ligand_target_matrix %>% 
  rownames() %>% 
  convert_human_to_mouse_symbols()

ligand_target_matrix <- ligand_target_matrix %>% 
    .[!is.na(rownames(ligand_target_matrix)), !is.na(colnames(ligand_target_matrix))]

weighted_network_lr <- weighted_network_lr %>% 
  mutate(from = convert_human_to_mouse_symbols(from),
         to = convert_human_to_mouse_symbols(to)) %>% 
    drop_na()
```

# Perform the NicheNet analysis  
In this case study, we want to apply NicheNet to predict which ligands expressed by all immune cells in the T cell area of the lymph node are most likely to have induced the differential expression in CD8 T cells after LCMV infection.  

the pipeline of a basic NicheNet analysis consist of the following steps:  

1. Define a 'sender/niche' cell population and a 'receiver/target' cell population present in your expression data and determine which genes are expressed in both populations.  

In this case study, the receiver cell population is the ‘CD8 T’ cell population, whereas the sender cell populations are ‘CD4 T,’ ‘Treg,’ ‘Mono,’ ‘NK,’ ‘B’ and ‘DC.’ We will consider a gene to be expressed when it is expressed in at least 10% of cells in one cluster.  

```{r}
# receiver 
receiver = "CD8 T"
expressed_genes_receiver <- get_expressed_genes(ident = receiver, seurat_obj = seuratObj, pct = 0.1)
background_expressed_genes <- expressed_genes_receiver %>% 
  .[. %in% rownames(ligand_target_matrix)]
```

```{r}
# sender
sender_celltypes <- c("CD4 T", "Treg", "Mono", "NK", "B", "DC")
list_expressed_genes_sender <- sender_celltypes %>% 
    unique() %>% 
    lapply(get_expressed_genes, seuratObj, pct = 0.1)

expressed_genes_sender <- list_expressed_genes_sender %>% 
    unlist() %>% 
    unique()
```

2. Define a gene set of interest: these are the genes in the "receiver/target" cell population that are potentially affected by ligands expressed by interacting cells. (e.g., genes differentially expressed upon cell-cell interaction).  

Here, the gene set of interest are the genes differentially expressed in CD8 T cells after LCMV infection. The condition of interest is thus ‘LCMV,’ whereas the reference/steady-state condition is ‘SS.’ The notion of conditions can be extracted from the metadata column ‘aggregate.’ The method to calculate the differential expression is here the standard Seurat Wilcoxon test, but this can be changed if necessary.  

```{r}
seurat_obj_receiver <- subset(seuratObj, idents = receiver)
seurat_obj_receiver <- SetIdent(seurat_obj_receiver, value = seurat_obj_receiver[["aggregate"]])

condition_oi <- "LCMV"
condition_reference = "SS"

DE_table_receiver <- FindMarkers(object = seurat_obj_receiver, 
                                 ident.1 = condition_oi, 
                                 ident.2 = condition_reference, 
                                 min.pct = 0.10) %>% 
    rownames_to_column("gene")

geneset_oi <- DE_table_receiver %>% 
    filter(p_val_adj <= 0.05, abs(avg_log2FC) >= 0.25) %>% 
    pull(gene)

geneset_oi <- geneset_oi %>% .[. %in% rownames(ligand_target_matrix)]
```

3. Define a set of potential ligands: these are ligands that are expressed by the 'sender/niche' cell population and bind a (putative) receptor expressed by the 'receiver/target' population.  

Because we combined the expressed genes of each sender cell type, in this example, we will perform one NicheNet analysis by pooling all ligands from all cell types together. Later on during the interpretation of the output, we will check which sender cell type expresses which ligand.  

```{r}
ligands <- lr_network %>% pull(from) %>% unique()
receptors <- lr_network %>% pull(to) %>% unique()

expressed_ligands <- intersect(ligands, expressed_genes_sender)
expressed_receptors <- intersect(receptors, expressed_genes_receiver)

potential_ligands <- lr_network %>% 
    filter(from %in% expressed_ligands, to %in% expressed_receptors) %>% 
    pull(from) %>% 
    unique()
```

4. Perform NicheNet ligand activity analysis: rank the potential ligands based on the presence of their target genes in the gene set of interest (compared to the background set of genes).  

```{r}
ligand_activities <- predict_ligand_activities(geneset = geneset_oi, 
                                               background_expressed_genes = background_expressed_genes, 
                                               ligand_target_matrix = ligand_target_matrix, 
                                               potential_ligands = potential_ligands)
ligand_activities %>% arrange(-pearson)
```

The different ligand activity measures (auroc, aupr, pearson correlation coefficient) are a measure for how well a ligand can predict the observed differentially expressed genes compared to the background of expressed genes. In our validation study, we showed that the pearson correlation coefficient between a ligand’s target predictions and the observed transcriptional response was the most informative measure to define ligand activity. Therefore, NicheNet ranks the ligands based on their pearson correlation coefficient. This allows us to prioritize ligands inducing the antiviral response in CD8 T cells.  

The number of top-ranked ligands that are further used to predict active target genes and construct an active ligand-receptor network is here 20.  

```{r}
best_upstream_ligands <- ligand_activities %>% 
    top_n(20, pearson) %>% 
    arrange(-pearson) %>% 
    pull(test_ligand) %>% 
    unique()
```

These ligands are expressed by one or more of the input sender cells. To see which cell population expresses which of these top-ranked ligands, you can run the following:

```{r}
DotPlot(object = seuratObj, 
        features = best_upstream_ligands, 
        cols = "RdYlBu") +
    RotatedAxis()
```

5. Infer receptors and top-predicted target genes of ligands that are top-ranked in the ligand activity analysis.  

**active target gene inference**  

```{r}
active_ligand_target_link_df <- best_upstream_ligands %>% 
    lapply(get_weighted_ligand_target_links, 
           geneset = geneset_oi, 
           ligand_target_matrix = ligand_target_matrix, 
           n = 200) %>% 
    bind_rows() %>% 
    drop_na()

active_ligand_target_links <- prepare_ligand_target_visualization(
    ligand_target_df = active_ligand_target_link_df, 
    ligand_target_matrix = ligand_target_matrix, 
    cutoff = 0.33
)

order_ligands = intersect(best_upstream_ligands, colnames(active_ligand_target_links)) %>% 
    rev() %>% 
    make.names()

order_targets = active_ligand_target_link_df$target %>% 
    unique() %>% 
    intersect(rownames(active_ligand_target_links)) %>% 
    make.names()

rownames(active_ligand_target_links) = rownames(active_ligand_target_links) %>% 
    make.names() # make.names() for heatmap visualization of genes like H2-T23

colnames(active_ligand_target_links) = colnames(active_ligand_target_links) %>% 
    make.names() # make.names() for heatmap visualization of genes like H2-T23

vis_ligand_target = active_ligand_target_links[order_targets,order_ligands] %>% t()
```

```{r}
p_ligand_target_network <- vis_ligand_target %>% 
    make_heatmap_ggplot(matrix = ., 
                        x_name = "Prioritized ligands", 
                        y_name = "Predicted target genes", 
                        color = "purple", 
                        legend_position = "top",
                        x_axis_position = "top", 
                        legend_title = "Regulatory potential") +
    theme(axis.text.x = element_text(face = "italic")) +
    scale_fill_gradient2(low = "whitesmoke", high = "purple", breaks = c(0, 0.0045, 0.0090))
```

Note that not all ligands from the top 20 are present in this ligand-target heatmap. The left-out ligands are ligands that don’t have target genes with high enough regulatory potential scores. Therefore, they did not survive the used cutoffs. To include them, you can be less stringent in the used cutoffs.  

**Receptors of top-ranked ligands**  

```{r}
lr_network_top <- lr_network %>% 
    filter(from %in% best_upstream_ligands, to %in% expressed_receptors) %>% 
    distinct(from, to)

best_upstream_receptors <- lr_network_top %>% 
    pull(to) %>% 
    unique()

lr_network_top_df_large <- weighted_network_lr %>% 
    filter(from %in% best_upstream_ligands, to %in% best_upstream_receptors)

lr_network_top_df <- lr_network_top_df_large %>% 
    spread("from", "weight", fill = 0)

lr_network_top_matrix <- lr_network_top_df %>% 
    select(-to) %>% 
    as.matrix() %>% 
    magrittr::set_rownames(lr_network_top_df$to)

dist_receptors <- dist(lr_network_top_matrix, method = "binary")
hclust_receptors <- hclust(dist_receptors, method = "ward.D2")
order_receptors <- hclust_receptors$labels[hclust_receptors$order]

dist_ligands <- dist(lr_network_top_matrix %>% t(), method = "binary")
hclust_ligands <- hclust(dist_ligands, method = "ward.D2")
order_ligands <- hclust_ligands$labels[hclust_ligands$order]

order_receptors <- order_receptors %>% intersect(rownames(lr_network_top_matrix))
order_ligands <- order_ligands %>% intersect(colnames(lr_network_top_matrix))

vis_ligand_receptor_network <- lr_network_top_matrix[order_receptors, order_ligands]
rownames(vis_ligand_receptor_network) <- order_receptors %>% make.names()
colnames(vis_ligand_receptor_network) <- order_ligands %>% make.names()
```

```{r}
p_ligand_receptor_network <- vis_ligand_receptor_network %>% 
    make_heatmap_ggplot(y_name = "ligands", 
                        x_name = "receptors", 
                        color = "mediumvioletred", 
                        x_axis_position = "top", 
                        legend_title = "Prior interaction potential", 
                        legend_position = "right") +
    theme(aspect.ratio = 0.618)

p_ligand_receptor_network
```

**Receptors of top-ranked ligands, but after considering only bona fide ligand-receptor interactions documented in literature and publicly available databases**  

```{r}
lr_network_strict = lr_network %>% filter(database != "ppi_prediction_go" & database != "ppi_prediction")
ligands_bona_fide = lr_network_strict %>% pull(from) %>% unique()
receptors_bona_fide = lr_network_strict %>% pull(to) %>% unique()

lr_network_top_df_large_strict = lr_network_top_df_large %>% distinct(from,to) %>% inner_join(lr_network_strict, by = c("from","to")) %>% distinct(from,to)
lr_network_top_df_large_strict = lr_network_top_df_large_strict %>% inner_join(lr_network_top_df_large, by = c("from","to"))

lr_network_top_df_strict = lr_network_top_df_large_strict %>% spread("from","weight",fill = 0)
lr_network_top_matrix_strict = lr_network_top_df_strict %>% select(-to) %>% as.matrix() %>% magrittr::set_rownames(lr_network_top_df_strict$to)

dist_receptors = dist(lr_network_top_matrix_strict, method = "binary")
hclust_receptors = hclust(dist_receptors, method = "ward.D2")
order_receptors = hclust_receptors$labels[hclust_receptors$order]

dist_ligands = dist(lr_network_top_matrix_strict %>% t(), method = "binary")
hclust_ligands = hclust(dist_ligands, method = "ward.D2")
order_ligands_receptor = hclust_ligands$labels[hclust_ligands$order]

order_receptors = order_receptors %>% intersect(rownames(lr_network_top_matrix_strict))
order_ligands_receptor = order_ligands_receptor %>% intersect(colnames(lr_network_top_matrix_strict))

vis_ligand_receptor_network_strict = lr_network_top_matrix_strict[order_receptors, order_ligands_receptor]
rownames(vis_ligand_receptor_network_strict) = order_receptors %>% make.names()
colnames(vis_ligand_receptor_network_strict) = order_ligands_receptor %>% make.names()
```

```{r}
p_ligand_receptor_network_strict = vis_ligand_receptor_network_strict %>% 
    t() %>% 
    make_heatmap_ggplot("Ligands",
                        "Receptors", 
                        color = "mediumvioletred", 
                        x_axis_position = "top",
                        legend_title = "Prior interaction potential\n(bona fide)")

p_ligand_receptor_network_strict
```

6. Add log fold change information of ligands from sender cells  

In some cases, it might be possible to also check upregulation of ligands in sender cells. This can add a useful extra layer of information next to the ligand activities defined by NicheNet, because you can assume that some of the ligands inducing DE in receiver cells, will be DE themselves in the sender cells.  

Here this is possible: we will define the log fold change between LCMV and steady-state in all sender cell types and visualize this as extra information.  

```{r}
# DE analysis for each sender cell type
DE_table_all <- Idents(seuratObj) %>% 
    levels() %>% 
    intersect(sender_celltypes) %>% 
    lapply(get_lfc_celltype, 
           seurat_obj = seuratObj, 
           condition_colname = "aggregate", 
           condition_oi = condition_oi, 
           condition_reference = condition_reference, 
           expression_pct = 0.10) %>% 
  reduce(full_join)

# combine ligand activities with DE information
ligand_activities_de <- ligand_activities %>% 
  select(test_ligand, pearson) %>% 
  rename(ligand = test_ligand) %>% 
  left_join(DE_table_all %>% rename(ligand = gene))

# make LFC heatmap 
lfc_matrix <- ligand_activities_de %>% 
  select(-ligand, -pearson) %>% 
  as.matrix() %>% 
  magrittr::set_rownames(ligand_activities_de$ligand)

rownames(lfc_matrix) <- rownames(lfc_matrix) %>% make.names()

order_ligands <- order_ligands[order_ligands %in% rownames(lfc_matrix)]
vis_ligand_lfc <- lfc_matrix[order_ligands, ]

colnames(vis_ligand_lfc) <- vis_ligand_lfc %>% colnames() %>% make.names()

p_ligand_lfc <- vis_ligand_lfc %>% 
  make_threecolor_heatmap_ggplot(y_name = "Prioritized ligands", 
                                 x_name = "LFC in Sender", 
                                 low_color = "white", 
                                 mid = median(vis_ligand_lfc), 
                                 mid_color = "white",
                                 high_color = "red", 
                                 legend_position = "top", 
                                 x_axis_position = "top", 
                                 legend_title = "LFC") + 
  theme(axis.text.y = element_text(face = "italic"))
```

7. Summary visulization of the NicheNet analysis  
For example, you can make a combined heatmap of ligand activities, ligand expression, ligand log fold change and the target genes of the top-ranked ligands. The plots for the log fold change and target genes were already made. Let’s now make the heatmap for ligand activities and for expression.  














