---
title: "model-construction"
author: "yincy"
date: "2/29/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This vignette shows how ligand-target prior regulatory potential scores are inferred in the NicheNet framework. You can use the procedure shown here to develop your own model with inclusion of context-specific networks or removal of noisy irrelevant data sources.  

## Background information about NicheNet's prior ligand-target model  
The prior model at the basis of NicheNet denotes how strongly existing knowledge supports that a ligand may regulate the expression of a target gene. To calculate this ligand-target regulatory potential, we integrated biological knowledge about ligand-to-target signaling paths as follows.  

First, we collected multiple complementary data sources covering ligand-receptor, signal transduction (e.g., protein-protein and kinase-substrate interactions) and gene regulatory interactions (e.g., inferred from ChIP-seq and motifs). For information of all collected data sources see [Data source information](https://github.com/saeyslab/nichenetr/blob/master/vignettes/data_sources.xlsx).  

Secondly, we integrated these individual sources into two weigted networks:  

1. a ligand-signaling network, which contains protein-protein interactions covering the signaling paths from ligands to downstream transcription regulators.  

2. a gene regulatory network, which contains gene regulatory interactions between transcriptional regulators and target geens.   

**To let informative data sources contribute more to the final model, we weighted each data source during integration**. These data source weights were automatically determined via model-based parameter optimization to improve the accuracy of ligand-target predictions. (see the vignette [Parameter optimization via mlrMBO](https://github.com/saeyslab/nichenetr/blob/master/vignettes/parameter_optimization.md).)  

In this vignette, we will show how to construct models with unoptimized data source weights as well.  

**Finally, we combined the ligand-signaling and gene regulatory network to calculate a regulatory potential score between all pairs of ligands and target genes. A ligand-target pair receives a high regulatory potential if the regulators of the target gene are lying downstream of the signaling network of the ligand**. 

To calculate this, we used network propagation methods on the integrated networks to propagate the signal starting from a ligand, flowing through receptors, signaling proteins, transcriptional regulators, and ultimately ending at target genes.  

```{r, fig.cap="A graphical summary of this procedure",}
knitr::include_graphics(path = "figures/workflow_model_construction.png")
```


## Construct a ligand-target model from all collected ligand-receptor, signaling and gene regulatory network data sources  
```{r, message=FALSE}
library(nichenetr)
library(tidyverse)
```

In the NicheNet framework, ligand-target links are predicted based on collected biological knowledge on ligand-receptor, signaling and gene regulatory interactions.  
```{r}
lr_network <- readRDS(file = "../../../Data/NicheNet/data/lr_network.rds")
# ligand receptor network

sig_network <- readRDS(file = "../../../Data/NicheNet/data/signaling_network.rds")
# receptor signaling network

gr_network <- readRDS(file = "../../../Data/NicheNet/data/gr_network.rds")
# gene regulatory network
```


### Construct NicheNet's ligand-target model from unoptimized data source weights  
Construct the weighted integrated ligand-signaling and gene regulatory network. In this first example, we give every data source the same weight (as given by the source_weights_df data frame provided by default by the nichenetr package).  

For the hyperparameters of the model (hub correction factors and damping factor), we will use the optimized values (as given by the `hyperparameter_list` data frame provided by default by the nichenetr package).  

The ligand-signaling network hub correction factor and gene regulatory network hub correction factor were defined as hyperparameter of the model to mitigate the potential negative influence of over-dominant hubs on the final model.  

The damping factor hyperparameter is the main parameter of the Personalized PageRank algorithm, which we used as network propagation algorithm to link ligands to downstream regulators.  

aggregate the individual data sources in a weighted manner to obtain a weighted integrated signaling network  
```{r}
weighted_networks <- construct_weighted_networks(lr_network = lr_network, 
                                                 sig_network = sig_network,
                                                 gr_network = gr_network, 
                                                 source_weights_df = source_weights_df)
```


downweight the importance of signaling and gene regulatory hubs - use the optimized parameters of this  
```{r}
weighted_networks <- apply_hub_corrections(weighted_networks = weighted_networks, 
                                           lr_sig_hub = hyperparameter_list$lr_sig_hub,
                                           gr_hub = hyperparameter_list$gr_hub)
```

Infer ligand-target regulatory potential scores based on the weighted integrated networks  
```{r}
# in this example we will calculate target gene regulatory potential scores for TNF and the ligand combination TNF+IL6  
ligands <- list("TNF", c("TNF", "IL6"))
ligand_target_matrix <- construct_ligand_target_matrix(weighted_networks = weighted_networks, 
                                                       ligands = ligands, 
                                                       algorithm = "PPR", 
                                                       damping_factor = hyperparameter_list$damping_factor, 
                                                       ltf_cutoff = hyperparameter_list$ltf_cutoff)
```

Show some top target genes of the ligand TNF and the ligand combination TNF + IL6  
```{r}
extract_top_n_targets(ligand_oi = "TNF", 
                      top_n = 10, 
                      ligand_target_matrix = ligand_target_matrix)
```

```{r}
extract_top_n_targets(ligand_oi = "TNF-IL6", 
                      top_n = 10, 
                      ligand_target_matrix = ligand_target_matrix)
```


### Construct NicheNet's ligand-target model from optimized data source weights  
```{r}
# aggregate the individual data sources in a weighted manner to obtain a weighted integrated signaling network
weighted_networks <- construct_weighted_networks(
    lr_network = lr_network, 
    sig_network = sig_network, 
    gr_network = gr_network, 
    source_weights_df = optimized_source_weights_df
)


# downweight the importance of signaling and gene regulatory hubs - use the optimized parameters of this
weighted_networks <- apply_hub_corrections(
    weighted_networks = weighted_networks, 
    lr_sig_hub = hyperparameter_list$lr_sig_hub, 
    gr_hub = hyperparameter_list$gr_hub
)


# Infer ligand-target regulatory potential scores based on the weighted integrated network
ligands = list("TNF")

ligand_target_matrix <- construct_ligand_target_matrix(
    weighted_networks = weighted_networks, 
    ligands = ligands, 
    algorithm = "PPR", 
    damping_factor = hyperparameter_list$damping_factor, 
    ltf_cutoff = hyperparameter_list$ltf_cutoff
)

ligand_target_matrix %>% head()
```


## Change the data sources at the basis of the NicheNet ligand-target model  
### Keep only specific data sources of interest  
Now, we will demonstrate how you can decide which data sources to use in the model you want to create. Let’s say for this example, that you are interested in making a model that only consists of literature-derived ligand-receptor interactions, signaling and gene regulatory interactions from comprehensive databases and gene regulatory interactions inferred from ChIP-seq. An annotation of the different data sources is given by the annotation_data_sources data frame provided by default by the nichenetr package).  
```{r}
annotation_data_sources %>% pull(type_db) %>% table()
```

```{r}
data_sources_to_keep <- annotation_data_sources %>% 
    filter(type_db %in% c("literature", "comprehensive_db", "ChIP")) %>% 
    pull(source)
```

```{r}
new_source_weights_df <- source_weights_df %>% 
    filter(source %in% data_sources_to_keep)
```

```{r}
new_lr_network <- lr_network %>% 
    filter(source %in% data_sources_to_keep)

new_gr_network <- gr_network %>%
    filter(source %in% data_sources_to_keep)

new_sig_network <- sig_network %>% 
    filter(source %in% data_sources_to_keep)
```


```{r}
weighted_network <- construct_weighted_networks(
    lr_network = new_lr_network, 
    sig_network = new_sig_network, 
    gr_network = new_gr_network, 
    source_weights_df = new_source_weights_df
)

weighted_network <- apply_hub_corrections(
    weighted_networks = weighted_network, 
    lr_sig_hub = hyperparameter_list$lr_sig_hub, 
    gr_hub = hyperparameter_list$gr_hub
)

ligands <- list("TNF")
ligand_target_matrix <- construct_ligand_target_matrix(
    weighted_networks = weighted_network, 
    ligands = ligands, 
    ltf_cutoff = hyperparameter_list$ltf_cutoff, 
    algorithm = "PPR", 
    damping_factor = hyperparameter_list$damping_factor
)
```

```{r}
extract_top_n_targets(ligand_oi = "TNF", top_n = 10, ligand_target_matrix = ligand_target_matrix)
```


### Add own data sources to the NicheNet model  
In addition to removing data sources, you can also add new data sources. This could for example help you in making context-specific models, if you would have a network or data containing context-specific interactions of interest.  

As input, we required a data source to contain directional interactions between genes: these interactions are protein-protein or signaling interactions for ligand-receptor and signaling data sources and a gene regulatory interaction for gene regulatory data sources. The data sources should be formatted in a data frame with following columns: from, to and source. “from” denotes the source node “gene A” of the directional interaction from gene A to B, “to” denotes the target node “gene B” of this directional interaction, and “source” is a user-defined name of this data source.  

[example data](these interactions are protein-protein or signaling interactions for ligand-receptor and signaling data sources and a gene regulatory interaction for gene regulatory data sources. The data sources should be formatted in a data frame with following columns: from, to and source. “from” denotes the source node “gene A” of the directional interaction from gene A to B, “to” denotes the target node “gene B” of this directional interaction, and “source” is a user-defined name of this data source.)  

```{r}
input_file <- "https://amp.pharm.mssm.edu/static/hdfs/harmonizome/data/hubs/gene_attribute_edges.txt.gz"
ppi_network <- read_tsv(file = input_file, 
                        col_names = TRUE)

ppi_network <- ppi_network %>% 
    transmute(from = target, to = source) %>% 
    filter(from %in% geneinfo_human$symbol & to %in% geneinfo_human$symbol)

# give your data source name
ppi_network <- ppi_network %>% 
    mutate(source = "harmonizome_hub_ppi", 
           database = "harmonizome")
```

First, we will add this new data source to all other data sources. **Because this data sources contains intracellular protein-protein interactions, we will consider this data source as a signaling data source**. As example, we will assign to this data source a weight of 1, because we want it to have a strong contribution to the final model.  
```{r}
new_sig_network <- sig_network %>% 
    bind_rows(ppi_network)
```


```{r}
new_network_weights_tf <- tibble(source = "harmonizome_hub_ppi", weight = 1)
new_source_weights_df <- optimized_source_weights_df %>% 
    bind_rows(new_network_weights_tf)
```


```{r}
weighted_network <- construct_weighted_networks(
    lr_network = lr_network, 
    sig_network = new_sig_network, 
    gr_network = gr_network, 
    source_weights_df = new_source_weights_df
)

weighted_network <- apply_hub_corrections(
    weighted_networks = weighted_network, 
    lr_sig_hub = hyperparameter_list$lr_sig_hub, 
    gr_hub = hyperparameter_list$gr_hub)

ligands <- list("TNF")
ligand_target_matrix <- construct_ligand_target_matrix(
    weighted_networks = weighted_network, 
    ligands = ligands, 
    ltf_cutoff = hyperparameter_list$ltf_cutoff, 
    damping_factor = hyperparameter_list$damping_factor, 
    algorithm = "PPR"
)
```

```{r}
extract_top_n_targets("TNF", 10, ligand_target_matrix)
```











