---
title: "Signac"
author: "yincy"
date: "3/6/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

For this tutorial, we will be analyzing a single-cell ATAC-seq dataset of human peripheral blood mononuclear cells (PBMCs) provided by 10x Genomics.  

```{r, message=FALSE, echo=FALSE}
library(Signac)
library(Seurat)
library(GenomeInfoDb)
library(tidyverse)
library(cowplot)
library(EnsDb.Hsapiens.v75)
```


## Quick start  
The vignette below presents a detailed workflow for the pre-processing, QC, clustering, visualization, integration, and interpretation of single-cell ATAC-seq data. While there are many steps (some of which are computationally intensive), we note that the core analytical components are similar to single-cell RNA-seq, and can be run quickly in a few lines. For example, the following lines encapsulate our scATAC-seq workflow (but does not include the calculation of advanced QC metrics, or integration with scRNA-seq data).  
```{r}
counts <- Read10X_h5(filename = "/home/yincy/git/Data/Signac/atac_v1_pbmc_10k_filtered_peak_bc_matrix.h5")
metadata <- read.csv(file = "/home/yincy/git/Data/Signac/atac_v1_pbmc_10k_singlecell.csv", 
                     header = T, 
                     row.names = 1)

pbmc_fast <- CreateSeuratObject(counts = counts, meta.data = metadata) %>% 
    subset(peak_region_fragments > 1000) %>% 
    RunTFIDF() %>% 
    FindTopFeatures(cutoff = "q75") %>% 
    RunSVD(reduction.name = "lsi") %>% 
    FindNeighbors(reduction = "lsi", dims = 1:30) %>% 
    FindClusters(resolution = 0.5, verbose = F) %>% 
    RunUMAP(reduction = "lsi", dims = 1:30)

DimPlot(pbmc_fast, label = F)
```

## Pre-processing workflow  
When pre-processing chromatin data, Signac uses information from two related input files, both of which are created by CellRanger:  
- **Peak/Cell matrix**. This is analogous to the gene expression count matrix used to analyze single-cell RNA-seq. However, instead of genes, each row of the matrix represents a region of the genome (a ‘peak’), that is predicted to represent a region of open chromatin. Each value in the matrix represents the number of Tn5 cut sites for each single barcode (i.e. cell) that map within each peak. You can find more detail on the [10X Website](https://support.10xgenomics.com/single-cell-atac/software/pipelines/latest/output/peaks).  

- **Fragment file**. This represents a full list of all unique fragments across all single cells. It is a substantially larger file, is slower to work with, and is stored on-disk (instead of in memory). However, the advantage of retaining this file is that it contains all fragments associated with each single cell, as opposed to only reads that map to peaks.

For most analyses we work with the peak/cell matrix, but for some (e.g. creating a ‘Gene Activity Matrix’), we find that restricting only to reads in peaks can adversely affect results. We therefore use both files in this vignette. We start by creating a Seurat object using the peak/cell matrix, and store the path to the fragment file on disk in the Seurat object:   
```{r}
counts <- Read10X_h5(filename = "/home/yincy/git/Data/Signac/atac_v1_pbmc_10k_filtered_peak_bc_matrix.h5")
metadata <- read.csv(file = "/home/yincy/git/Data/Signac/atac_v1_pbmc_10k_singlecell.csv", 
                     header = T, 
                     row.names = 1) %>% .[-1, ]

pbmc <- CreateSeuratObject(counts = counts, 
                           assay = "peaks", 
                           project = "ATAC", 
                           min.cells = 0, 
                           min.features = 0,
                           meta.data = metadata)

fragment.path <- "/home/yincy/git/Data/Signac/atac_v1_pbmc_10k_fragments.tsv.gz"
pbmc <- SetFragments(
    object = pbmc, 
    file = fragment.path,
    assay = "peaks"
)
```

## Filtering the fragment file (Optional)  
To make downstream steps that use this file faster, we can filter the fragments file to contain only reads from cells that we retain in the analysis. This is optional, but slow, and only needs to be performed once. Running this command writes a new file to disk and indexes the file so it is ready to be used by Signac.  
```{r}
fragment_file_filtered <- "/home/yincy/git/Data/Signac/atac_v1_pbmc_10k_filtered_fragments.tsv"

FilterFragments(
    fragment.path = fragment.path, 
    cells = colnames(pbmc), 
    output.path = fragment_file_filtered
)

pbmc <- SetFragments(object = pbmc, file = paste0(fragment_file_filtered, ".bgz"))
```


## Computing QC Metrics  
We can now compute some QC metrics for the scATAC-seq experiment. We currently suggest the following metrics below to assess data quality. As with scRNA-seq, **the expected range of values for these parameters will vary depending on your biological system, cell viability, etc**.  

- **Nucleosome banding pattern**: The histogram of fragment sizes (determined from the paired-end sequencing reads) should exhibit a strong nucleosome banding pattern. We calculate this per single cell, and quantify the approximate ratio of mononucleosomal to nucleosome-free fragments (stored as nucleosome_signal). Note that by default, this is calculated only on chr1 reads (see the `region` parameter) to save time.(recomended: nucleosome signal > 10)   

- **Transcriptional start site (TSS) enrichment score**. The ENCODE project has defined an ATAC-seq targeting score based on the ratio of fragments centered at the TSS to fragments in TSS-flanking regions (see [https://www.encodeproject.org/data-standards/terms/](https://www.encodeproject.org/data-standards/terms/)). Poor ATAC-seq experiments typically will have a low TSS enrichment score. We can compute this metric for each cell with the `TSSEnrichment` function, and the results are stored in metadata under the column name `TSS.enrichment`.(TSS.enrichment > 2)  

Transcription start site (TSS) enrichment value are dependent on the reference files used: cutoff values for high quality data are listed in the table below.  

**Annotation used**              | **Value**   | **Resulting Data Status**
---------------------------------|-------------|------------------------------
hg19 Refseq TSS annotation       | < 6         | Concerning
                                 | 6 - 10      | Acceptable
                                 | > 10        | Ideal
GRCh38 Refseq annotation         | < 5         | Concerning
                                 | 5 - 7       | Acceptable
                                 | > 7         | Ideal
mm9 GENCODE TSS annotation       | < 5         | Concerning
                                 | 5 - 7       | Acceptable 
                                 | > 7         | Ideal
mm10 Refseq TSS annotation       | < 10        | Concerning
                                 | 10 - 15     | Acceptable
                                 | > 15        | Ideal


- **Total number of fragments in peaks**: A measure of cellular sequencing depth/complexity. Cells with very few reads may need to be excluded due to low sequencing depth. Cells with extremely high levels may represent doublets or nuclear clumps.(3000 < ragments_in_peaks < 20000)  

- **Fraction of fragments in peaks**: Represents the fraction of total fragments that fall within ATAC-seq peaks. Cells with low values (i.e. < 15 - 20%) often represent low-quality cells or technical artifacts that should be removed.(should be > 0.3 but >0.2 are acceptable)  

- **Ratio reads in "blacklist" sits**: The ENCODE project has provided a list of [blacklist regions](https://github.com/Boyle-Lab/Blacklist), representing reads which are often associated with artifactual signal. Cells with a high proportion of reads mapping to these areas (compared to reads mapping to peaks) often represent technical artifacts and should be removed. ENCODE blacklist regions for human (hg19 and GRCh38), mouse (mm10), Drosophila (dm3), and C.elemegts(ce10) are included in the Signac package.(< 5%)  

- The alignment rate, or percentage of mapped reads, should be greater than 95%, though values > 80% may be acceptable.  

Note that the last three metrics can be obtained from the output of CellRanger (which is stored in the object metadata), but can also be calculated for non-10x datasets using Signac (more information at the end of this document).  

```{r}
pbmc <- NucleosomeSignal(object = pbmc, assay = "peaks")
pbmc$pct_reads_in_peaks <- pbmc$peak_region_fragments / pbmc$passed_filters * 100
pbmc$blacklist_ratio <- pbmc$blacklist_region_fragments / pbmc$peak_region_fragments
```


```{r plot QC}
p1 <- pbmc[[]] %>%
    ggplot() +
    geom_violin(aes(x = "ATAC", y = pct_reads_in_peaks), fill = "tomato", alpha = 0.7) +
    geom_jitter(aes(x = "ATAC", y = pct_reads_in_peaks), width = 0.4, size = 0.1, alpha = 1/3) +
    scale_x_discrete(name = NULL, breaks = NULL) +
    scale_y_continuous(name = NULL) +
    ggtitle(label = "pct_reads_in_peaks") +
    theme(panel.background = element_rect(fill = "white"), 
          axis.line = element_line(color = "black"))

p2 <- pbmc[[]] %>% 
    ggplot()+
    geom_violin(aes(x = "ATAC", y = blacklist_ratio + 0.001), fill = "tomato", alpha = 0.7) +
    geom_jitter(aes(x = "ATAC", y = blacklist_ratio + 0.001), width = 0.4, size = 0.1, alpha = 1/3) +
    scale_x_discrete(name = NULL, breaks = NULL) +
    scale_y_continuous(name = NULL) +
    ggtitle(label = "blacklist_ratio") +
    theme(panel.background = element_rect(fill = "white"), 
          axis.line = element_line(color = "black"))

p3 <- pbmc[[]] %>% 
    ggplot()+
    geom_violin(aes(x = "ATAC", y = nucleosome_signal + 0.001), fill = "tomato", alpha = 0.7) +
    geom_jitter(aes(x = "ATAC", y = nucleosome_signal + 0.001), width = 0.4, size = 0.1, alpha = 1/3) +
    scale_x_discrete(name = NULL, breaks = NULL) +
    scale_y_continuous(name = NULL) +
    ggtitle(label = "nucleosome_signal") +
    theme(panel.background = element_rect(fill = "white"), 
          axis.line = element_line(color = "black"))

p4 <- pbmc[[]] %>% 
    ggplot()+
    geom_violin(aes("ATAC", peak_region_fragments + 0.001), fill = "tomato", alpha = 0.7) +
    geom_jitter(aes("ATAC", peak_region_fragments + 0.001), width = 0.4, size = 0.1, alpha = 1/3) +
    scale_x_discrete(name = NULL, breaks = NULL) +
    scale_y_continuous(name = NULL) +
    ggtitle(label = "peak_region_fragments") +
    theme(panel.background = element_rect(fill = "white"), 
          axis.line = element_line(color = "black"))

plot_grid(plotlist = list(p1, p2, p3, p4), nrow = 2, align = "hv")
```

```{r, message=FALSE, warning=FALSE}
pbmc[["nucleosome_group"]] <- ifelse(pbmc[["nucleosome_signal"]] > 10, "NS > 10", "NS < 10")
FragmentHistogram(object = pbmc, group.by = "nucleosome_group", region = genes[seqnames(genes) == "chr2"]) +
    geom_vline(xintercept = c(147, 294), linetype = 2, lwd = 0.4, color = "grey70")
# the length of mononucleosome is 147 - 294 bp
```

The enrichment of Tn5 integration events at transcriptional start sites (TSSs) can also be an important quality control metric to assess the targeting of Tn5 in ATAC-seq experiments. The ENCODE consortium defined a TSS enrichment score as the number of Tn5 integration site around the TSS normalized to the number of Tn5 integration sites in flanking regions.  

See the ENCODE documentation for more information about the TSS enrichment score (https://www.encodeproject.org/data-standards/terms/). We can calculate the TSS enrichment score for each cell using the TSSEnrichment function in Signac.  
```{r}
# create Granges object with TSS positions  
gene.ranges <- genes(EnsDb.Hsapiens.v75)
gene.ranges <- gene.ranges[gene.ranges$gene_biotype == "protein_coding", ]

tss.ranges <- GRanges(
    seqnames = seqnames(gene.ranges), 
    ranges = IRanges(start = start(gene.ranges), width = 2), 
    strand = strand(gene.ranges)
)

seqlevelsStyle(tss.ranges) <- "UCSC"
tss.ranges <- keepStandardChromosomes(x = tss.ranges, pruning.mode = "coarse")
pbmc <- TSSEnrichment(object = pbmc, tss.positions = tss.ranges[1:2000])
```


```{r}
pbmc$high.tss <- ifelse(pbmc$TSS.enrichment > 2, "high", "low")
TSSPlot(object = pbmc, group.by = "high.tss") + ggtitle("TSS enrichment score") + NoLegend()
```


Finally remove cells that are outliers for these QC metrics.   
```{r}
pbmc <- subset(pbmc, subset = peak_region_fragments > 3000 & 
                   peak_region_fragments < 20000 & 
                   pct_reads_in_peaks > 15 & 
                   blacklist_ratio < 0.05 &  
                   nucleosome_signal < 10 & 
                   TSS.enrichment > 2)
```


## Normalization and linear dimensional reduction  

- `Normalization`: Signac performs term frequency-inverse document frequency (TF-IDF) normalization. This is a two-step normalization procedure, that both normalizes across cells to correct for differences in cellular sequencing depth, and across peaks to give higher values to more rare peaks.   

- `Feature selection`: The largely binary nature of scATAC-seq data makes it challenging to perform ‘variable’ feature selection, as we do for scRNA-seq. Instead, we can choose to use only the top n% of features (peaks) for dimensional reduction, or remove features present in less that n cells with the FindTopFeatures function. Here, we will all features, though we note that we see very similar results when using only a subset of features (try setting min.cutoff to ‘q75’ to use the top 25% all peaks), with faster runtimes. Features used for dimensional reduction are automatically set as VariableFeatures for the Seurat object by this function.  

- `Dimensional reduction`: We next run a singular value decomposition (SVD) on the TD-IDF normalized matrix, using the features (peaks) selected above. This returns a low-dimensional representation of the object (for users who are more familiar with scRNA-seq, you can think of this as analogous to the output of PCA).  

```{r}
pbmc <- RunTFIDF(object = pbmc, assay = "peaks", method = 1) %>% 
    FindTopFeatures(min.cutoff = "q0") %>% 
    RunSVD(assay = "peaks", reduction.key = "LSI_", reduction.name = "lsi", irlba.work = 500)
```


### Non-linear dimension reduction and clustering  
Now that the cells are embedded in a low-dimensional space, we can use methods commonly applied for the analysis of scRNA-seq data to perform graph-based clustering, and non-linear dimension reduction for visualization. The functions `RunUMAP`, `FindNeighbors`, and `FindClusters` all come from the Seurat package.  
```{r}
pbmc <- RunUMAP(object = pbmc, reduction = "lsi", dims = 2:30) %>% 
    FindNeighbors(reduction = "lsi", dims = 2:30) %>% 
    FindClusters(verbose = F, algorithm = 3)

DimPlot(object = pbmc, label = T) + NoLegend()
```


### Create a gene activity matrix  
The UMAP visualization reveals the presence of multiple cell groups in human blood. If you are familiar with our scRNA-seq analyses of PBMC, you may even recognize the presence of certain myeloid and lymphoid populations in the scATAC-seq data. However, annotating and interpreting clusters is more challenging in scATAC-seq data, as much less is known about the functional roles of noncoding genomic regions than is known about protein coding regions (genes).  

However, we can try to quantify the activity of each gene in the genome by assessing the chromatin accessibility associated with each gene, and create a new gene activity assay derived from the scATAC-seq data. Here we will use a simple approach of summing the reads intersecting the gene body and promoter region (we also recommend exploring the Cicero tool, which can accomplish a similar goal).   

To create a gene activity matrix, we extract gene coordinates for the human genome from EnsembleDB, and extend them to include the 2kb upstream region (as promoter accessibility is often correlated with gene expression). We then count the number of fragments for each cell that map to each of these regions, using the `FeatureMatrix` function. This takes any set of genomic coordinates, counts the number of reads intersecting these coordinates in each cell, and returns a sparse matrix.  

```{r}
# extract gene coordinates from Ensemble, and ensure name formatting is consistence with Seurat object
gene.coords <- genes(x  =EnsDb.Hsapiens.v75, filter = ~ gene_biotype == "protein_coding")
seqlevelsStyle(gene.coords) <- "UCSC"
genebody.coords <- keepStandardChromosomes(x = gene.coords, pruning.mode = "coarse")
genebodyandPromotor.coords <- Extend(x = genebody.coords, upstream = 2000, downstream = 0)

# create a gene by cell matrix  
gene.activities <- FeatureMatrix(
    fragments = fragment.path, 
    features = genebodyandPromotor.coords, 
    cells = colnames(pbmc), 
    chunk = 20
)

# convert rownames from chromosomal coordinates into gene names
gene.key <- genebodyandPromotor.coords$gene_name
names(gene.key) <- GRangesToString(grange = genebodyandPromotor.coords)
rownames(gene.activities) <- gene.key[rownames(gene.activities)]
# for that you should known how the peak fragment is assigned
```

```{r}
# add the gene activity matrix to the Seurat object as a new assay, and normalize it
pbmc[["RNA"]] <- CreateAssayObject(counts = gene.activities)
pbmc <- NormalizeData(
    object = pbmc, 
    assay = "RNA", 
    normalization.method = "LogNormalize", 
    scale.factor = median(pbmc$nCount_RNA)
)
```

Now, we can visualize the ‘activities’ of canonical marker genes to help interpret our ATAC-seq clusters. Note that the activities will be much noisier than scRNA-seq measurements. This is because they represent predictions from sparse chromatin data, but also because they assume a general correspondence between gene body/promoter accessibility and expression, which may not always be the case. Nonetheless, we can begin to discern populations of monocytes, B, T, and NK cells. However, further subdivision of these cell types is challenging based on supervised analysis alone.  
```{r}
DefaultAssay(pbmc) <- "RNA"

FeaturePlot(
    object = pbmc, 
    features = c("MS4A1", "CD3D", "LEF1", "NKG7", "TREM1", "LYZ"), 
    pt.size = 0.1, 
    max.cutoff = "q95", 
    ncol = 3
)
```


### Integrating with scRNA-seq data  
To help interpret the scATAC-seq data, we can classify cells based on an scRNA-seq experiment from the same biological system (human PBMC). We utilize methods for cross-modality integration and label transfer, described here, with a more in-depth tutorial here. We aim to identify shared correlation patterns in the gene activity matrix and scRNA-seq dataset in order to identify matched biological states across the two modalities. This procedure returns a classification score for each cell, based on each scRNA-seq defined cluster label.  
```{r}
pbmc_rna <- readRDS(file = "/home/yincy/git/Data/Seurat/scATAC--scRNA-seq/pbmc_10k_v3.rds")

transfer.anchors <- FindTransferAnchors(
    reference = pbmc_rna, 
    query = pbmc, 
    reduction = "cca"
)

predicted.labels <- TransferData(anchorset = transfer.anchors, 
                                 refdata = pbmc_rna$celltype, 
                                 weight.reduction = pbmc[["lsi"]])
```

```{r}
predicted.labels[, -1] %>% 
    apply(1, max) %>% 
    hist(breaks = 50, col = "orange", border = NA)
axis(side = 1, at = seq(1.0, 0.2, by = -0.1))
abline(v = c(predicted.labels[, -1] %>% apply(1, max) %>% quantile(probs = c(0.1, 0.2)), 0.5), lty = 2, lwd = 2, col  ="red")
```

```{r}
pbmc <- AddMetaData(object = pbmc, metadata = predicted.labels)
```


```{r}
plot1 <- DimPlot(object = pbmc_rna, 
                 group.by = "celltype", 
                 label = TRUE, 
                 repel = T) +
    NoLegend() +
    ggtitle("scRNA-seq")

plot2 <- DimPlot(object = pbmc, 
                 group.by = "predicted.id", 
                 label = TRUE, 
                 repel = T) +
    NoLegend() +
    ggtitle("scATAC-seq")

plot_grid(plot1, plot2)
```

```{r}
pbmc <- subset(pbmc, idents = 14, invert  =TRUE)
pbmc <- RenameIdents(object = pbmc, 
                     "0" = "CD14 Mono", 
                     "1" = "CD4 Memory", 
                     "2" = "CD8 Effector", 
                     "3" = "CD4 Naive", 
                     "4" = "CD14 Mono", 
                     "5" = "CD8 Naive", 
                     "6" = "DN T", 
                     "7" = "NK CD56Dim", 
                     "8" = "pre-B", 
                     "9" = "CD16 Mono", 
                     "10" = "pro-B", 
                     "11" = "DC", 
                     "12" = "NK CD56bright", 
                     "13" = "pDC")

```


### Find differentially accessible peaks between clusters  
To find differentially accessible regions between clusters of cells, we can perform a differential accessibility (DA) test. We utilize logistic regression for DA, as suggested by Ntranos et al. 2018 for scRNA-seq data, and add the total number of fragments as a latent variable to mitigate the effect of differential sequencing depth on the result.  
```{r}
# switch back working with peaks instead of gene activities  
DefaultAssay(pbmc) <- "peaks"

da_peaks <- FindMarkers(object = pbmc, 
                        ident.1 = "CD4 Naive", 
                        ident.2 = "CD14 Mono", 
                        min.pct = 0.2, 
                        test.use = "LR", 
                        latent.vars = "peak_region_fragments")
da_peaks %>% head()
```

```{r}
da_p1 <- VlnPlot(object = pbmc, 
                 features = rownames(da_peaks)[1], 
                 ncol = 3, 
                 pt.size = 0.1,
                 idents = c("CD4 Memory", "CD14 Mono"))

da_p2 <- FeaturePlot(object = pbmc, 
                 features = rownames(da_peaks)[1], 
                 ncol = 3, 
                 pt.size = 0.1)
plot_grid(da_p1, da_p2)
```

Peak coordinates can be difficult to interpret alone. We can find the closest gene to each of these peaks using the ClosestFeature function and providing an EnsDb annotation. If you explore the gene lists, you will see that peaks open in Naive T cells are close to genes such as BCL11B and GATA3 (key regulator of T cell differentiation), while peaks open in monocytes are close to genes such as CEBPB (key regulator of monocyte differentiation). We could follow up this result further by doing gene ontology enrichment analysis on the gene sets returned by ClosestFeature, and there are many R packages that can do this.  
```{r}
open_cd4naive <- rownames(da_peaks[da_peaks$avg_logFC > 0.5, ])
open_cd14mono <- rownames(da_peaks[da_peaks$avg_logFC < -0.5, ])

closest_genes_cd4naive <- ClosestFeature(regions = open_cd4naive, 
                                         annotation = EnsDb.Hsapiens.v75, 
                                         sep = c(':', "-"))

closest_gene_cd14mono <- ClosestFeature(regions = open_cd14mono, 
                                        annotation = EnsDb.Hsapiens.v75, 
                                        sep = c(":", "-"))
```

We can also create coverage plots grouped by cluster around any genomic region using the CoveragePlot function. These represent ‘pseudo-bulk’ accessibility tracks, where all cells within a cluster have been averaged together, in order to visualize a more robust chromatin landscape.  
```{r, message=FALSE, warning=FALSE}
# set plot order
levels(pbmc) <- c("CD4 Naive","CD4 Memory","CD8 Naive","CD8 Effector","DN T","NK CD56bright","NK CD56Dim","pre-B",'pro-B',"pDC","DC","CD14 Mono",'CD16 Mono')

cp <- CoveragePlot(object = pbmc, 
             region = rownames(da_peaks)[c(1, 5)], 
             sep = c(":", "-"), 
             peaks = StringToGRanges(regions = rownames(pbmc), sep = c(":", "-")), 
             annotation = EnsDb.Hsapiens.v75, 
             extend.upstream = 20000, 
             extend.downstream = 20000, 
             ncol = 2)
cp
```


## Working with datasets that were not quantified using CellRanger  
The CellRanger software from 10x Genomics generates several useful QC metrics per-cell, as well as a peak/cell matrix and an indexed fragments file. In the above vignette, we utilize the CellRanger outputs, but provide alternative functions in Signac for many of the same purposes here.  

### Generating a peak/cell or bin/cell matrix  
The `FeatureMatrix` function can be used to generate a count matrix containing any set of genomic ranges in its rows. These regions could be a set of peaks, or bins that span the entire genome.  
```{r}
# peak_ranges should be a set of genomic ranges spaning the set of peaks to be quantified per cell
peak_matrix <- FeatureMatrix(
    fragments = fragment.path, 
    features = gene_ranges, 
    sep = c(":", "-")
)
```

For convenience, we also include a GenomeBinMatrix function that will generate a set of genomic ranges spanning the entire genome for you, and run FeatureMatrix internally to produce a genome bin/cell matrix.  
```{r}
bin_matrix <- GenomeBinMatrix(
    fragments = fragment.path, 
    genome = "hg19", 
    binsize = 5000
)
```


### Counting fraction of reads in peaks  
The function FRiP will count the fraction of reads in peaks for each cell, given a peak/cell assay and a bin/cell assay. Note that this can be run on a subset of the genome, so that a bin/cell assay does not need to be computed for the whole genome. This will return a Seurat object will metadata added corresponding to the fraction of reads in peaks for each cell.  
```{r}
pbmc <- FRiP(
    object = pbmc, 
    peak.assay = "peaks", 
    bin.assay = "bins"
)
```


### Counting fragments in genome blacklist regions  
The ratio of reads in genomic blacklist regions, that are known to artifactually accumulate reads in genome sequencing assays, can be diagnostic of low-quality cells. We provide blacklist region coordinates for several genomes (hg19, hg38, mm9, mm10, ce10, ce11, dm3, dm6) in the Signac package for convenience. These regions were provided by the ENCODE consortium, and we encourage users to cite their [paper](https://doi.org/10.1038/s41598-019-45839-z) if you use the regions in your analysis. The `FractionCountsInRegion` function can be used to calculate the fraction of all counts within a given set of regions per cell. We can use this function and the blacklist regions to find the fraction of blacklist counts per cell.  

```{r}
pbmc$blacklist_ratio <- FractionCountsInRegion(
    object = pbmc, 
    assay = "peaks", 
    regions = blacklist_hg19
)
```






















