---
title: "ChIP-seq"
author: "yincy"
date: "12/4/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = F, warning = F)
```


## Workflow of ChIP-seq Analysis  
- Quality Assessment of sequencing: FastQC  

- Mapping: bowtie1/2, Rsubread  

- Peak calling: CCAT, SICER, MACS, ZINBA, BayesPeak, chipseq, ChIPseqR, CSAR, csaw, GenoGAM, iSeq, PICS  

- Quality assessment:ChIPQC  

- Peak annotation and pathway analysis: ChIPpeakAnno, Homer, PAVIS, GREAT, ChIPseeker, clusterProfiler, GOstats  

- Gene Network Building: GeneNetworkBuilder, ChIPXpress  

- Visualization: trackViewer, IGV, UCSC genome browser, rtracklayer  

- Motif enrichment analysis: The MEME Suit, homer  


```{r}
library(GEOquery)
library(SRAdb)
```


```{r}
sqlfile <- "SRAmetadb.sqlite"
if(!file.exists(sqlfile))
    sqlfile <- getSRAdbFile()

sra_con <- dbConnect(SQLite(), sqlfile)
conversion <- sraConvert(in_acc = c("SRP055170"), 
                         sra_con = sra_con)

out <- getSRAfile(conversion$sample,
                  sra_con = sra_con,
                  fileType = "sra")
```





