---
title: "SCopeLoomR"
author: "YinCY"
date: "6/1/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction
The .loom file format is designed to store very large omics datasets. It has been created is maintained by the **Linnarsson lab**, and its naming conventions are described in `loompy`.

`SCopeLoomR` is an R package to easily create and manipulate `.loom` files. These loom files are compatible with `SCope`, a loom viewer which also allows to visualize the results from `SCENIC` (e.g.: regulon, regulon activities).  

# Creating a loom object
## load data
```{r}
library(SingleCellExperiment)
library(SCopeLoomR)
library(magrittr)

data(sce)

# DGEM (digital gene expression matrix)
dgem <- counts(sce)

dgem %>% dim
dgem %>% class
```

```{r}
colnames(dgem) %>% head
```


```{r}
cell.info <- colData(sce)
cell.info$nGene <- colSums(dgem > 0)

cell.info %>% head
```

```{r}
# default embedding (e.g. t-SNE or PCA coordinates)
data("tSNE_expr")
default.tsne <- tSNE_expr
default.tsne.name <- "t-SNE on full expression matrix"

default.tsne %>% head
```

# Create loom file
Minimum information required to create the .loom object

```{r}
library(SCopeLoomR)

# create the minimal loom file
file.name <- "example.loom"

build_loom(
  file.name = file.name, 
  dgem = dgem, 
  title = "fake expression dataset for examples", 
  genome = "mm10", 
  default.embedding = default.tsne, 
  default.embedding.name = default.tsne.name
)
```

# Add optional information
To add any following information to a loom, please run the following command before

```{r}
loom <- open_loom(file.path = file.name, mode = "r+")
```

## Hierarchy
You can organize/group your .loom files in SCope by specifying differents grouping levels. The current .loom file will be put in `Mouse -> Toy Datasets` branch of the SCope loom tree.  

```{r}
add_hierarchy(loom = loom, 
              hierarchy = create_hierarchy(
                level.1.name = "Mouse", 
                level.2.name = "Toy Datasets", 
                level.3.name = ""
              ))
```


The same command can be used to update the hierarchy of a .loom file (set overwrite=TRUE):
```{r}
add_hierarchy(loom = loom, 
              hierarchy = create_hierarchy(
                level.1.name = "[level-1-name]", 
                level.2.name = "[level-2-name]", 
                level.3.name = "[level-3-name]"
              ), 
              overwrite = T)
```


## Annotations/metrics
Annotations and/or metrics can be added to query in `SCope`.  

```{r}
# add annotation (categorical variable)
add_col_attr(
  loom = loom, 
  key = "cell type", 
  value = cell.info$cellType, 
  as.annotation = T
)

# add metric (numerical variable)
add_col_attr(
  loom = loom, 
  key = "Age", 
  value = sample(0:20, nrow(cell.info), replace = T), 
  as.metric = T
)
```


# SCENIC results
```{r}
scenic.dir <- file.path(system.file("extdata", package = "SCopeLoomR"), "SCENIC_fakeOutput/")

# regulon activity (SUC matrix)
library(AUCell)

regulonsAUC <- readRDS(file.path(scenic.dir, "int/3.4_regulonAUC.Rds"))
add_scenic_regulons_auc_matrix(loom = loom, regulons.AUC = getAUC(regulonsAUC))

# regulons (gene list), regulon thresholds (optional) and regulon motifs (optional)
regulons <- readRDS(file.path(scenic.dir, "int/3.1_regulons_forAUCell.Rds"))
aucThresholds <- readRDS(file.path(scenic.dir, "int/3.5_AUCellThresholds.Rds"))
regulon.enrichment.table <- readRDS(file = "2.3_motifEnrichment.Rds")

add_scenic_regulons(loom = loom, 
                    regulons = regulons, 
                    regulon.threshold.assignments = aucThresholds, 
                    regulon.enrichment.table = regulon.enrichment.table)

# alternative t-SNE
tSNE <- readRDS(file.path(scenic.dir, "int/tSNE_AUC_50pcs_50perpl.Rds"))
add_embedding(loom = loom, 
              embedding = tSNE$Y, 
              name = "SCENIC (t-SNE on AUC)")
```

```{r}
finalize(loom = loom)
```


# Read data from a loom object
## Get the gene expression matrix
```{r}
loom_path <- "example.loom"
loom <- open_loom(loom_path, mode = "r+")
```

```{r}
dgem <- get_dgem(loom = loom)
close_loom(loom = loom)

dgem[1:5, 1:5]
```






