---
title: "Chapter 13 Integrating Datasets"
author: "YinCY"
date: "6/7/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Motivation
Large single-cell RNA sequencing (scRNA-seq) projects usually need to generate data across multiple batches due to logistical constraints. However, the processing of different batches is often subject to uncontrollable differences, e.g., changes in operator, differences in reagent quality. This results in systematic differences in the observed expression in cells from different batches, which we refer to as “batch effects”. Batch effects are problematic as they can be major drivers of heterogeneity in the data, masking the relevant biological differences and complicating interpretation of the results.  

Computational correction of these effects is critical for eliminating batch-to-batch variation, allowing data across multiple batches to be combined for common downstream analysis. However, existing methods based on linear models (Ritchie et al. 2015; Leek et al. 2012) assume that the composition of cell populations are either known or the same across batches. To overcome these limitations, bespoke methods have been developed for batch correction of single-cell data (Haghverdi et al. 2018; Butler et al. 2018; Lin et al. 2019) that do not require a priori knowledge about the composition of the population. This allows them to be used in workflows for exploratory analyses of scRNA-seq data where such knowledge is usually unavailable.  

## Setting up the data
To demonstrate, we will use two separate 10X Genomics PBMC datasets generated in two different batches. Each dataset was obtained from the `TENxPBMCData` package and separately subjected to basic processing steps. Separate processing prior to the batch correction step is more convenient, scalable and (on occasion) more reliable. For example, outlier-based QC on the cells is more effective when performed within a batch (Section 6.3.2.3). The same can also be said for trend fitting when modelling the mean-variance relationship (Section 8.2.4.1).  

```{r}
library(TENxPBMCData)

all.sce <- list(
  pbmc3k = TENxPBMCData(dataset = "pbmc3k"), 
  pbmc4k = TENxPBMCData(dataset = "pbmc4k"), 
  pbmc8k = TENxPBMCData(dataset = "pbmc8k")
)

# quality control
library(scater)
stats <- high.mito <- list()

for(n in names(all.sce)){
    current = all.sce[[n]]
    is.mito = grep("^MT", rowData(current)$Symbol_TENx)
    stats[[n]] = perCellQCMetrics(current, subsets = list(Mito = is.mito))
    high.mito[[n]] = isOutlier(stats[[n]]$subsets_Mito_percent, type = "higher")
    all.sce[[n]] = current[, !high.mito[[n]]]
}

# normalization
all.sce <- lapply(all.sce, logNormCounts)

# variance modelling
library(scran)
all.dec <- lapply(all.sce, modelGeneVar)
all.hvgs <- lapply(all.dec, getTopHVGs, prop = 0.1)

# dimensionality reduction
library(BiocSingular)
set.seed(10000)

all.sce <- mapply(FUN = runPCA, 
                  x = all.sce, 
                  subset_row = all.hvgs, 
                  MoreArgs = list(ncomponents = 25, 
                                  BSPARAM = RandomParam()), 
                  SIMPLIFY = FALSE)

set.seed(100000)
all.sce <- lapply(all.sce, runTSNE, dimred = "PCA")

set.seed(1000000)
all.sce <- lapply(all.sce, runUMAP, dimred = "PCA")

# clustering
for(n in names(all.sce)){
  g = buildSNNGraph(all.sce[[n]], k = 10, use.dimred = "PCA")
  clust = igraph::cluster_walktrap(g)$membership
  colLabels(all.sce[[n]]) = factor(clust)
}
```

```{r}
pbmc3k <- all.sce$pbmc3k
pbmc4k <- all.sce$pbmc4k
pbmc8k <- all.sce$pbmc8k

dec3k <- all.dec$pbmc3k
dec4k <- all.dec$pbmc4k
dec8k <- all.dec$pbmc8k
```

To prepare for the batch correction:  

1. We subset all batches to the common "universe" of features. In this case, it is straightforward as both batches use Ensembl gene annotation.  

```{r}
universe <- intersect(rownames(pbmc3k), rownames(pbmc4k))
length(universe)

pbmc3k <- pbmc3k[universe, ]
pbmc4k <- pbmc4k[universe, ]

dec3k <- dec3k[universe, ]
dec4k <- dec4k[universe, ]
```

2. We rescale each batch to adjust for differences in sequencing depth between batches. The `multiBatchNorm()` function recomputes log-normalized expression values after adjusting the size factors for systematic differences in coverage between `SingleCellExperiment` objects. (Size factors only remove biases between cells within a single batch.) This improves the quality of the correction by removing one aspect of the technical differences between batches.  

```{r}
library(batchelor)

rescaled <- multiBatchNorm(pbmc3k, pbmc4k)
pbmc3k <- rescaled[[1]]
pbmc4k <- rescaled[[2]]
```

3. We perform feature selection by averaging the variance components across all batches with the `combineVar()` function. We compute the average as it is responsive to batch-specific HVGs while still preserving the within-batch ranking of genes. This allows us to use the same strategies described in Section 8.3 to select genes of interest. In contrast, approaches based on taking the intersection or union of HVGs across batches become increasingly conservative or liberal, respectively, with an increasing number of batches.  

```{r}
library(scran)
combined.dec <- combineVar(dec3k, dec4k)
chosen.hvgs <- combined.dec$bio > 0
sum(chosen.hvgs)
```

**When integrating datasets of variable composition, it is generally safer to err on the side of including more genes than are used in a single dataset analysis, to ensure that markers are retained for any dataset-specific subpopulations that might be present**. For a top X selection, this means using a larger X (say, ~5000), or in this case, we simply take all genes above the trend. That said, many of the signal-to-noise considerations described in Section 8.3 still apply here, so some experimentation may be necessary for best results.  

Alternatively, a more forceful approach to feature selection can be used based on marker genes from within-batch comparisons; this is discussed in more detail in Section 13.7.  

## Diagnosing batch effects
Before we actually perform any correction, it is worth examining whether there is any batch effect in this dataset. We combine the two SingleCellExperiments and perform a PCA on the log-expression values for all genes with positive (average) biological components. In this example, our datasets are file-backed and so we instruct runPCA() to use randomized PCA for greater efficiency - see Section 23.2.2 for more details - though the default IRLBA will suffice for more common in-memory representations.  

```{r}
# synchronizing the metadata for cbind
rowData(pbmc3k) <- rowData(pbmc4k)
pbmc3k$batch <- "3k"
pbmc4k$batch <- "4k"
uncorrected <- cbind(pbmc3k, pbmc4k)

# using RandomParam() as it is more efficient for file-backed matrices
library(scater)
set.seed(0010101010)
uncorrected <- runPCA(uncorrected, subset_row = chosen.hvgs, 
                      BSPARAM = BiocSingular::RandomParam())
```

We use graph-based clustering on the components to obtain a summary of the population structure. As our two PBMC populations should be replicates, each cluster should ideally consist of cells from both batches. However, we instead see clusters that are comprised of cells from a single batch. This indicates that cells of the same type are artificially separated due to technical differences between batches.  

```{r}
library(scran)
snn.gr <- buildSNNGraph(uncorrected, use.dimred = "PCA")
clusters <- igraph::cluster_walktrap(snn.gr)$membership
tab <- table(Cluster = clusters, Batch = uncorrected$batch)
```

We can also visualize the corrected coordinates using a t-SNE plot. The strong separation between cells from different batches is consistent with the clustering results.  

```{r}
set.seed(1111001)
uncorrected <- runTSNE(uncorrected, dimred = "PCA")
plotTSNE(uncorrected, color_by = "batch")
```

Of course, the other explanation for batch-specific clusters is that there are cell types that are unique to each batch. The degree of intermingling of cells from different batches is not an effective diagnostic when the batches involved might actually contain unique cell subpopulations (which is not a consideration in the PBMC dataset, but the same cannot be said in general). If a cluster only contains cells from a single batch, one can always debate whether that is caused by a failure of the correction method or if there is truly a batch-specific subpopulation. For example, do batch-specific metabolic or differentiation states represent distinct subpopulations? Or should they be merged together? We will not attempt to answer this here, only noting that each batch correction algorithm will make different (and possibly inappropriate) decisions on what constitutes “shared” and “unique” populations.  

## Linear regression









