---
title: "rtracklayer"
author: "yincy"
date: "11/10/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Steps  
```
session <- browserSession() # create a BrowserSession object
track(session, trackName = "") <- targetTrack # layer the track
view <- browserView(object = session, 
                    range = `GRanges`, # to focus a particular range
                    pack = "targets")
```


# Introduction  

The `rtracklayer` package is an interface (or layer ) between R and genome browsers. Its main purpose is the visualization of genomic annotation tracks, whether generated through experimental data analysis performed in R or loaded from an external data source.  

The features of **rtracklayer** may be divided into two categories:  

1. the import/export of track data  
2. the control and querying of external genome browser sessions and views.  

The basic track data structure in Bioconductor is the *GRanges* class, defined in the *GenomicRanges* package.  

**Rtracklayer** supports the import and export of tracks from and to files in various formats.  

1. Browser Extended Display (BED).  
2. General Feature Format (GFF).  
3. Wiggle (WIG).  

The **rtracklayer** package currently interfaces with the **UCSC** web-based genome browser.  


# Gene expression and microRNA target sites  
## Constructing the GRanges 

```{r}
library("humanStemCell")
data(fhesc)

library(genefilter)
filtFhesc <- nsFilter(fhesc)[[1]]

library(limma)
design <- model.matrix(~ filtFhesc$Diff)
hesclim <- lmFit(filtFhesc, design)
hesceb <- eBayes(hesclim)
tab <- topTable(hesceb, coef = 2, adjust.method = "BH", n = 7676)
tab2 <- tab[(tab$logFC > 1) & (tab$adj.P.Val < 0.01), ]
affyIDs <- rownames(tab2)

library(microRNA)
data("hsTargets")

library(hgu133plus2.db)
entrezIDs <- mappedRkeys(hgu133plus2ENTREZID[affyIDs])

library(org.Hs.eg.db)
mappedEntrezIDs <- entrezIDs[entrezIDs %in% mappedkeys(org.Hs.egENSEMBL)]
ensemblIDs <- mappedRkeys(org.Hs.egENSEMBLTRANS[mappedEntrezIDs])
targetMatches <- match(ensemblIDs, hsTargets$target, 0)

# same as data(targets)
targets <- hsTargets[targetMatches, ]
targets$chrom <- paste("chr", targets$chrom, sep = "")
```

creates track from the *targets* dataset.  

```{r}
library(rtracklayer)
library(GenomicRanges)

head(targets)
```

```{r}
targetRanges <- IRanges(targets$start, targets$end)
targetTrack <- with(targets, GRangesForUCSCGenome(genome = "hg18", 
                                                  chrom = chrom, 
                                                  ranges = targetRanges, 
                                                  strand = strand, 
                                                  name = name, 
                                                  target = target))

genome(targetTrack) %>% head()
seqlengths(targetTrack) %>% head()
```

## Accessing track information  
```{r}
seqnames(targetTrack) %>% head()

start(targetTrack) %>% head()

width(targetTrack) %>% head()

strand(targetTrack) %>% head()
```


## Subsetting a GRanges  
```{r}
targetTrack[1:10]

targetTrack[strand(targetTrack) == "+"]

targetTrack[seqnames(targetTrack) == "chr1"]
```

## Exporting and importing tracks  
```{r}
export(object = targetTrack, con = "targets.bed")

retrack <- import(con = "targets.bed")
```


```{r}
export(targetTrack, con = "targets.gff")
```


# Viewing the targets in a genome browser  
### Starting a session  
The first step towards interfacing with a browser is to start a browser session, represented in R as a BrowserSession object. A BrowserSession is primarily a container of tracks and genomic views.  

```{r}
session <- browserSession(object = "UCSC")

genomeBrowsers()
```

### Laying the track  
before a track can be viewed on the genome, it must be loaded into the session using the **track** function.  

```{r}
track(object = session, "targets") <- targetTrack
```


### Viewing the track  
```{r}
subTargetsTrack <- targetTrack[1]
view <- browserView(object = session, range = subTargetsTrack * -10, pack = "targets")
```


If multiple ranges are provided, multiple views are lanuched  
```{r}
view <- browserView(session, range = targetTrack[1:5] * -10, pack = "targets")
```


### A shortcut  
```{r}
browseGenome(object = targetTrack, range = subTargetsTrack * -10)
```


# Downloading Tracks from your Web Browser  
```{r}
loaded_tracks <- trackNames(session)

loaded_tracks %>% head()
```

```{r}
subTargetsTrack <- track(session, "targets")
```

















