---
title: "GEOmetadb"
author: "yincy"
date: "3/27/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r,message=FALSE, warning=FALSE}
library(GEOmetadb)
library(DBI)
library(RSQLite)
library(tidyverse)
```


```{r}
con <- dbConnect(drv = RSQLite::SQLite(), "f:/git/Data/GEOmetadb.sqlite")
```

```{r}
dbListFields(con, "gpl")
```

```{r}
dbReadTable(con, 'gsm') %>% 
  filter()
```


```{r}
hsa_ngs <- dbGetQuery(conn = con, 
           statement = "select gse.gse, gsm.title, gse.summary, 
           gse.pubmed_id, gpl.gpl, gpl.technology, gsm.gsm, 
           gsm.source_name_ch1, gsm.organism_ch1, gsm.molecule_ch1
           from gsm 
           join gse_gsm on gsm.gsm=gse_gsm.gsm 
           join gse on gse_gsm.gse=gse.gse 
           join gse_gpl on gse_gpl.gse=gse.gse 
           join gpl on gse_gpl.gpl=gpl.gpl 
           where gsm.organism_ch1 = 'Homo sapiens' 
           and technology = 'high-throughput sequencing'")
``` 


```{r}
hsa_ngs %>% 
    dplyr::filter(grepl("(AMD)", summary, ignore.case = F))
```


```{r}
hsa_retina_ngs %>% 
    filter((grepl("pbmc|`whole blood`", title) | grepl("pbmc|`whole blood`", summary) | grepl("pbmc|`whole blood`", source_name_ch1)) & technology == "high-throughput sequencing") %>% 
    pull(summary) %>% 
    unique()
```


AMD datasets: 
+ GSE99248: RPE-Choriod
+ GSE135092: late stage of amd is not specified  
+ GSE125564: RPE expression by array   
+ GSE146887: neovascular membrane in AMD patients (newly generated invision vascular)  
+ GSE29801: expression by array, RPEChoroid, retina  
+ SRP107997 (GSE99287): ATAC-seq of AMD RPE show decresed chromation accessibility  


```{r}
GSE99248_sens <- read.table(file = "/home/yincy/git/Data/AMD-related/GSE99248/GSE99248_master_list_of_gene_counts_MIN.sense.AMD_RNASeq.txt.gz", 
                            sep = "\t", 
                            header = T)

GSE99248_anti <- read.table(file = "/home/yincy/git/Data/AMD-related/GSE99248/GSE99248_master_list_of_gene_counts_MIN.antisense.AMD_RNASeq.txt.gz", 
                            sep = "\t", 
                            header = T)

GSE99248_sens %>% dim()
GSE99248_sens %>% .[1:5, 1:5]
GSE99248_anti %>% 
  select(id, geneSymbol, geneCoordinate, TR.18, TR.20, TR.32, TR02, TR06, TR08, TR10, TR12, TR14, TR16) %>% 
  filter(geneSymbol %in% amd)
```

```{r}
GSE99248_sens %>% 
  select(id, geneSymbol, geneCoordinate, TR.18, TR.20, TR.32, TR02, TR06, TR08, TR10, TR12, TR14, TR16) %>% 
  filter(geneSymbol %in% amd)
```


```{r}
library(GEOquery)
GSE99248_meta <- getGEO(GEO = "GSE99248")

RPE <- GSE99248_meta[[1]] %>% as.data.frame() %>% 
  dplyr::filter(grepl("Normal|Early", eye.histological.phenotype.ch1), grepl("RPE", source_name_ch1)) %>% 
  dplyr::pull(title)
```





