---
title: "PCAtools"
author: "yincy"
date: "4/9/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction  
Principal Component Analysis (PCA) is a very powerful technique that has wide applicability in data science, bioinformatics, and further afield. It was initially developed to analyse large volumes of data in order to tease out the differences/relationships between the logical entities being analysed. It extracts the fundamental structure of the data without the need to build any model to represent it. This ‘summary’ of the data is arrived at through a process of reduction that can transform the large number of variables into a lesser number that are uncorrelated (i.e. the ‘principal components’), while at the same time being capable of easy interpretation on the original data (Blighe and Lun 2019) (Blighe 2013).  

`PCAtools` provides functions for data exploration via PCA, and allows the user to generate publication-ready figures. PCA is performed via `BiocSingular` (Lun 2019) - users can also identify optimal number of principal components via different metrics, such as elbow method and Horn’s parallel analysis (Horn 1965) (Buja and Eyuboglu 1992), which has relevance for data reduction in single-cell RNA-seq (scRNA-seq) and high dimensional mass cytometry data.  

# Quick start  
For this vignette, we will load breast cancer gene expression data with recurrence free survival (RFS) from Gene Expression Profiling in Breast Cancer: Understanding the Molecular Basis of Histologic Grade To Improve Prognosis.  

```{r, message=FALSE, warning=FALSE}
library(Biobase)
library(GEOquery)
library(PCAtools)
```

```{r}
# load series and platform data from GEO
gset <- getGEO(GEO = "GSE2990", GSEMatrix = T, getGPL = F)
mat <- exprs(gset[[1]])

# remove Affymetrix control probs
mat <- mat[-grep("AFFX", rownames(mat)), ]

# extract information of interest from the phenotype data (pdata)
idx <- which(colnames(pData(gset[[1]])) %in% c("relation", "age:ch1", "distant rfs:ch1", 
                                               "er:ch1", "ggi:ch1", "grade:ch1", 
                                               "size:ch1", "time rfs:ch1"))
metadata <- data.frame(pData(gset[[1]])[, idx], 
                       row.names = rownames(pData(gset[[1]])))

# tidy column names
colnames(metadata) <- c("Study", "Age", "Distant.RFS", 
                        "ER", "GGI", "Grade", 
                        "Size", "Time.RFS")

# prepare certain phenotypes of interest  
metadata$Study <- gsub("Reanalyzed by: ", "", as.character(metadata$Study))
metadata$Age <- as.numeric(gsub("^KJ", NA, as.character(metadata$Age)))
metadata$Distant.RFS <- factor(metadata$Distant.RFS, levels = c(0, 1))
metadata$ER <- factor(gsub("\\?", NA, as.character(metadata$ER)), levels = c(0, 1))
metadata$ER <- factor(ifelse(metadata$ER == 1, "ER+", "ER-"), levels = c("ER-", "ER+"))
metadata$GGI <- as.numeric(as.character(metadata$GGI))
metadata$Grade <- gsub(1, "Grade 1", gsub(2, "Grade 2", gsub(3, "Grade 3", metadata$Grade)))
metadata$Grade <- factor(metadata$Grade, levels = c("Grade 1", "Grade 2", "Grade 3"))
metadata$Size <- as.numeric(as.character(metadata$Size))
metadata$Time.RFS <- as.numeric(gsub("^KJX|^KJ", NA, metadata$Time.RFS))

# remove samples from the pdata that have any NA value
discard <- apply(metadata, 1, function(x) any(is.na(x)))
metadata <- metadata[!discard, ]
# metadata <- na.omit(metadata)


# filter the expression data to match the samples in our pdata
mat <- mat[, which(colnames(mat) %in% rownames(metadata))]

# check that sample name smatch exactly between pdata and expression data
all(colnames(mat) == rownames(metadata))
```

## Conduct principal component analysis (PCA)
```{r}
p <- pca(mat, metadata = metadata, removeVar = 0.1)
```

## A scree plot
```{r}
screeplot(p, axisLabSize = 5, titleLabSize = 20)
```

## A bi-plot  






























