---
title: "ImmG"
author: "yincy"
date: "11/28/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = F, warning = F)
```


```{r}
library(GEOquery)
library(magrittr)
library(tidyverse)
```


Immune cell populations encompassing all major lineages were isolated by flow cytometry to high-purity according to ImmGen SOP [ImmGen](https://www.immgen.org/). 10,000 cells (except some rare cell populations) were used to construct each ATAC-seq library, using the Fast-ATAC protocol. Libraries were dual-indexed and pooled for high-throughput sequencing.
Highly purified cell samples were isolated in parallel from the same populations for gene expression profiling by low-input RNAseq (ImmGen protocol, per umbrella GSE126267). These corresponding RNAseq datasets are available at GSE109125.  


```{r}
gse100738_sample_info <- getGEO(GEO = "GSE100738", GSEMatrix = F)
```


```{r}
gse100738_header <- slot(gse100738_sample_info, "header")
gse100738_gsms <- slot(gse100738_sample_info, "gsms")
```


```{r}
gse100738_gsms %>% names() %>% head()
```


```{r}
gse100738_gsms %>% class()
```


```{r}
cells <- sapply(X = gse100738_gsms, FUN = function(x){
    cell = c(x@header$geo_accession, 
             x@header$source_name_ch1, 
             x@header$title, 
             x@header$characteristics_ch1[1],
             x@header$characteristics_ch1[2], 
             x@header$description[1], 
             x@header$description[2], 
             x@header$instrument_model)
}, simplify = TRUE)


cells <- t(cells) %>% as_tibble()
colnames(cells) <- c("geo_accession", "Source_name", "sample", "strain", "cell_type", "selection_marker", "supplement_file_extension", "instrument_model")
cells$strain <- gsub("strain: ", "", cells$strain)
cells$cell_type <- gsub("cell type: ", "", cells$cell_type)
cells$species <- rep("Mmusculus", dim(cells)[1])
```

```{r}
cells$Source_name_ch1 %>% unique()
```







