---
title: "ligand_activity_analysis"
author: "yincy"
date: "2/8/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## NicheNet's ligand activity analysis on a gene set of interest: predict active ligands and their target genes.  

In this vignette, you can learn how to perform a basic NicheNet analysis. A NicheNet analysis can help you to generate hypotheses about an intercellular communication process of interest for which you have bulk or single-cell gene expression data.  

Specifically, NicheNet can predict:  

  - which ligands from one cell population ('sender/niche') are most likely to affect target gene expression in an interacting cell population ('reciver/target).  
  - which specific target genes are affected by which of these predicted ligands.  
  
Because NicheNet studies how ligands affect gene expression in neighboring cells, you need to have data about this effect in gene expression you want to study. So, you need to have a clear set of genes that are putatively affected by ligands from one of more interacting cells.  

The pipline of a basic NicheNet analysis consist mainly of the following steps:  

  - Define a 'sender/niche' cell population and a 'receiver/target' cell population present in your expression data and determine which genes are expressed in both populations.  
  - Define a gene set of interest: these are the genes in the 'receiver/target' cell population that are potentially affected by ligands expressed by interacting cells (e.g. genes differentially expressed upon cell-cell interaction).  
  - Define a set of potential ligands: these are ligands that are expressed by the 'sender/target' population.  
  - Perform NichNet ligand activity analysis: rank the potential ligands based the presence of their target genes in the gene set of interest (compared to the background set of genes).  
  - Infer top-predicted target genes of ligands that are top-ranked in the ligand activity analysis.  
  
More specifically, we will look at which ligands expressed by cancer-associated fibroblasts (CAFs) can induce a specific gene program in neighboring malignant cells.  


## Step 0: Load required packages, NicheNet's ligand-target prior model and processed expression data of interacting cells  
```{r, message=FALSE, warning=FALSE}
library(nichenetr)
library(tidyverse)
library(magrittr)
```

Ligand-target model:  
This model denotes the prior potential that a particular ligand regulate the expression of a specific target gene.  

```{r}
ligand_target_matrix <- readRDS(file = 'data/ligand_target_matrix.rds')
ligand_target_matrix[1:5, 1:5] # target genes in rows, ligands in columns
```

Expression data of interacting cells: publicly available single-cell data from CAF and malignant cells from HNSCC tumors.  

```{r}
hnscc_expression <- readRDS(file = 'data/hnscc_expression.rds')

expression <- hnscc_expression$expression
sample_info <- hnscc_expression$sample_info
```
  
## Step 1: Define expressed genes in sender and receiver cell populations  
Our research question is to prioritize which ligands expressed by CAFs can induce p-EMT in neighboring malignant cells. Therefore, CAFs are the sender cells in this example and malignant cells are the receiver cells. This is an example of paracrine signaling. **Note that autocrine signaling can be considered if sender and receiver cell type are the same**.  

Now, we will determine which genes are expressed in the sender cells (CAFs) and receiver cells (malignant cells) from high quality primary tumors. Therefore, we wil not consider cells from tumor samples of less quality or from lymph node metastases.  

To determine expressed genes in this case study, we use the definition used by Puram et al. (the authors of this dataset), which is: Ea, the aggregate expression of each gene i across the k cells, calculated as Ea(i) = log2(average(TPM(i)1…k)+1), should be >= 4. We recommend users to define expressed genes in the way that they consider to be most appropriate for their dataset. For single-cell data generated by the 10x platform in our lab, we don’t use the definition used here, but we consider genes to be expressed in a cell type when they have non-zero values in at least 10% of the cells from that cell type. This is described as well in the other vignette [Perform NicheNet analysis starting from a Seurat object: step by step analysis](https://github.com/saeyslab/nichenetr/blob/master/vignettes/seurat_steps.md)  

```{r}
tumors_remove <- c("HN10","HN","HN12", "HN13", "HN24", "HN7", "HN8","HN23")

CAF_ids <- sample_info %>% 
  filter(`Lymph node` == 0 & !(tumor %in% tumors_remove) & `non-cancer cell type` == "CAF") %>% 
  pull(cell)

maligant_ids <- sample_info %>% 
  filter(`Lymph node` == 0 & !(tumor %in% tumors_remove) & `classified  as cancer cell` == 1) %>% 
  pull(cell)

expressed_genes_sender <- expression[CAF_ids, ] %>% 
  apply(2, function(x){10*(2**x - 1)}) %>% 
  apply(2, function(x){log2(mean(x) + 1)}) %>% 
  .[. >= 4] %>% 
  names()

expressed_gene_receiver <- expression[maligant_ids, ] %>% 
  apply(2, function(x){10*(2**x - 1)}) %>% 
  apply(2, function(x){log2(mean(x) + 1)}) %>% 
  .[. >= 4] %>% 
  names()

# Check the number of expressed genes: should be a "reasonable" number of total expressed genes in a cell type, e.g. between 5000 - 10000.  

length(expressed_genes_sender); length(expressed_gene_receiver)
```




























