---
title: "ligand_activity_analysis"
author: "yincy"
date: "2/8/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## NicheNet's ligand activity analysis on a gene set of interest: predict active ligands and their target genes.  

In this vignette, you can learn how to perform a basic NicheNet analysis. A NicheNet analysis can help you to generate hypotheses about an intercellular communication process of interest for which you have bulk or single-cell gene expression data.  

Specifically, NicheNet can predict:  

  - which ligands from one cell population ('sender/niche') are most likely to affect target gene expression in an interacting cell population ('reciver/target).  
  - which specific target genes are affected by which of these predicted ligands.  
  
Because NicheNet studies how ligands affect gene expression in neighboring cells, you need to have data about this effect in gene expression you want to study. So, you need to have a clear set of genes that are putatively affected by ligands from one of more interacting cells.  

The pipline of a basic NicheNet analysis consist mainly of the following steps:  

  - Define a 'sender/niche' cell population and a 'receiver/target' cell population present in your expression data and determine which genes are expressed in both populations.  
  - Define a gene set of interest: these are the genes in the 'receiver/target' cell population that are potentially affected by ligands expressed by interacting cells (e.g. genes differentially expressed upon cell-cell interaction).  
  - Define a set of potential ligands: these are ligands that are expressed by the 'sender/target' population.  
  - Perform NichNet ligand activity analysis: rank the potential ligands based the presence of their target genes in the gene set of interest (compared to the background set of genes).  
  - Infer top-predicted target genes of ligands that are top-ranked in the ligand activity analysis.  
  
More specifically, we will look at which ligands expressed by cancer-associated fibroblasts (CAFs) can induce a specific gene program in neighboring malignant cells.  


## Step 0: Load required packages, NicheNet's ligand-target prior model and processed expression data of interacting cells  
```{r, message=FALSE, warning=FALSE}
library(nichenetr)
library(tidyverse)
```

**Ligand-target model**:  
This model denotes the prior potential that a particular ligand regulate the expression of a specific target gene.  
```{r}
ligand_target_matrix <- readRDS(file = '../../../Data/NicheNet/data/ligand_target_matrix.rds')
ligand_target_matrix %>% dim()
ligand_target_matrix[1:5, 1:5] # target in rows, ligands in columns
```


Expression data of interacting cells: publicly available single-cell data from CAF and malignant cells from HNSCC tumors.  
```{r}
hnscc_expression <- readRDS(file = '../../../Data/NicheNet/data/hnscc_expression.rds')

expression <- hnscc_expression$expression
expression %>% dim()
expression %>% .[1:5, 1:5]

sample_info <- hnscc_expression$sample_info
sample_info %>% dim()
sample_info %>% .[1:5, 1:5]
```


## Step 1: Define expressed genes in sender and receiver cell populations  
Our research question is to prioritize which ligands expressed by CAFs can induce p-EMT in neighboring malignant cells. Therefore, CAFs are the sender cells in this example and malignant cells are the receiver cells. This is an example of paracrine signaling. **Note that autocrine signaling can be considered if sender and receiver cell type are the same**.  

Now, we will determine which genes are expressed in the sender cells (CAFs) and receiver cells (malignant cells) from high quality primary tumors. Therefore, we will not consider cells from tumor samples of less quality or from lymph node metastases.  

To determine expressed genes in this case study, we use the definition used by Puram et al. (the authors of this dataset), which is: Ea, the aggregate expression of each gene i across the k cells, calculated as Ea(i) = log2(average(TPM(i)~1…k~)+1), should be >= 4. We recommend users to define expressed genes in the way that they consider to be most appropriate for their dataset. For single-cell data generated by the 10x platform in our lab, we don’t use the definition used here, but **we consider genes to be expressed in a cell type when they have non-zero values in at least 10% of the cells from that cell type**. This is described as well in the other vignette [Perform NicheNet analysis starting from a Seurat object: step by step analysis](https://github.com/saeyslab/nichenetr/blob/master/vignettes/seurat_steps.md)  
```{r}
tumors_remove <- c("HN10","HN","HN12", "HN13", "HN24", "HN7", "HN8","HN23")

CAF_ids <- sample_info %>% 
  filter(`Lymph node` == 0 & !(tumor %in% tumors_remove) & `non-cancer cell type` == "CAF") %>% 
  pull(cell)

maligant_ids <- sample_info %>% 
  filter(`Lymph node` == 0 & !(tumor %in% tumors_remove) & `classified  as cancer cell` == 1) %>% 
  pull(cell)

expressed_genes_sender <- expression[CAF_ids, ] %>%  
  apply(2, function(x){10*(2**x - 1)}) %>% 
  apply(2, function(x){log2(mean(x) + 1)}) %>% 
  .[. >= 4] %>% 
  names()

expressed_gene_receiver <- expression[maligant_ids, ] %>% 
  apply(2, function(x){10*(2**x - 1)}) %>% 
  apply(2, function(x){log2(mean(x) + 1)}) %>% 
  .[. >= 4] %>% 
  names()

# Check the number of expressed genes: should be a "reasonable" number of total expressed genes in a cell type, e.g. between 5000 - 10000.  
length(expressed_genes_sender); length(expressed_gene_receiver)
```

## Step 2: Define the gene set of interest and a background of genes  
As gene set of interest, we consider the genes of which the expression is possibly affected due to communication with other cells. **The definition of this gene set depends on your research question and is a crucial step in the use of NicheNet**.  

Because we here want to investigate how CAFs regulate the expression of p-EMT genes in malignant cells, we will use the p-EMT gene set defined by Puram et al. as gene set of interest and use all genes expressed in malignant cells as background of genes.  
```{r}
geneset_oi <- read_tsv(file = "../../../Data/NicheNet/data/pemt_signature.txt", col_names = "genes") %>% 
  pull(genes) %>% 
  .[. %in% rownames(ligand_target_matrix)]

# only consider genes also present in the NicheNet model - this excludes genes from the gene list for which the official HGNC symbol was not used by Puram et al.  

background_expressed_genes <-  expressed_gene_receiver %>% 
  .[. %in% rownames(ligand_target_matrix)]

background_expressed_genes %>% head()
```


## Step 3: Define a set of potential ligands  
As potentially active ligands, we will use ligands that are:  

- expressed by CAFs   
- can bind a (putative) receptor expressed by malignant cells.  

Putative ligand-receptor links were gathered from NicheNet's ligand-receptor data sources.  

```{r}
lr_network <- readRDS(file = "../../../Data/NicheNet/data/lr_network.rds")

ligands <- lr_network %>% pull(from) %>% unique()
expressed_ligands <- intersect(x = ligands, y = expressed_genes_sender)

receptors <- lr_network %>% pull(to) %>% unique()
expressed_receptors <- intersect(x = receptors, y = expressed_gene_receiver)

lr_network_expressed <- lr_network %>% 
  filter(from %in% expressed_ligands & to %in% expressed_receptors)

lr_network_expressed
```

This ligand-receptor network contains the expressed ligand-receptor interactions. As potentially active ligands for the NicheNet analysis, we will consider the ligands from this network.  
```{r}
potential_ligands <- lr_network_expressed %>% 
  pull(from) %>% 
  unique()

potential_ligands %>% head()
```


## Step 4: Perform NicheNet's ligand activity analysis on the gene set of interest  
Now perform the ligand activity analysis: in this analysis, we will calculate the ligand activity of each ligand, or in other words, we will assess how well each CAF-ligand can predict the p-EMT gene set compared to the background of expressed genes (predict whether a gene belongs to the p-EMT program or not).  

```{r}
ligand_activities <- predict_ligand_activities(geneset = geneset_oi,
                                               background_expressed_genes = background_expressed_genes,
                                               ligand_target_matrix = ligand_target_matrix, 
                                               # regulatory potential scores between ligands and targets
                                               potential_ligands = potential_ligands)

ligand_activities %>% 
  arrange(desc(pearson))
```

Now, we want to rank the ligands based on their ligand activity. **In our validation study, we showed that the pearson correlation coefficient (PCC) between a ligand’s target predictions and the observed transcriptional response was the most informative measure to define ligand activity**. Therefore, we will rank the ligands based on their pearson correlation coefficient. This allows us to prioritize p-EMT-regulating ligands.  
```{r}
ligand_activities %>% 
  pull(pearson) %>% 
  hist(breaks = seq(min(ligand_activities$pearson), max(ligand_activities$pearson), length.out = 30),
       main = "The distribution of Pearson value", 
       col = 'orange', 
       border = "grey")

abline(v = ligand_activities %>% arrange(desc(pearson)) %>% 
         pull(pearson) %>% .[20], col = "red", lty = 2, lwd = 2)
```


```{r}
best_upstream_ligands <- ligand_activities %>% 
  top_n(n = 20, wt = pearson) %>% 
  arrange(desc(pearson)) %>% 
  pull(test_ligand)

best_upstream_ligands %>% head()
```


We see here that the performance metrics indicate that the 20 top-ranked ligands can predict the p-EMT genes reasonably, this implies that ranking of the ligands might be accurate as shown in our study. However, it is possible that for some gene sets, the target gene prediction performance of the top-ranked ligands would not be much better than random prediction. In that case, prioritization of ligands will be less trustworthy.  

Additional note: we looked at the top 20 ligands here and will continue the analysis by inferring p-EMT target genes of these 20 ligands. However, the choice of looking only at the 20 top-ranked ligands for further biological interpretation is based on biological intuition and is quite arbitrary. Therefore, users can decide to continue the analysis with a different number of ligands. We recommend to check the selected cutoff by looking at the distribution of the ligand activity values.  


## Step 5: Infer target genes of top-ranked ligands and visualize in a heatmap  
Now we will show how you can look at the regulatory potential scores between ligands and target genes of interest. In this case, we will look at links between top-ranked p-EMT regulating ligands and p-EMT genes. In the ligand-target heatmaps, we show here regulatory potential scores for interactions between the 20 top-ranked ligands and following target genes: genes that belong to the gene set of interest and to the 250 most strongly predicted targets of at least one of the 20 top-ranked ligands (the top 250 targets according to the general prior model, so not the top 250 targets for this dataset). Consequently, genes of your gene set that are not a top target gene of one of the prioritized ligands, will not be shown on the heatmap.  

```{r}
active_ligand_target_links_df <- best_upstream_ligands %>% 
  lapply(FUN = get_weighted_ligand_target_links, 
         geneset = geneset_oi, 
         ligand_target_matrix = ligand_target_matrix, 
         n = 250) %>% 
  bind_rows()

active_ligand_target_links_df %>% dim()

active_ligand_target_links_df %>% 
  arrange(-weight)
```

```{r}
active_ligand_target_links_df %>% 
  pull(weight) %>% 
  hist(breaks = seq(min(active_ligand_target_links_df$weight), max(active_ligand_target_links_df$weight), length.out = 60),
       main = "Distribution of Weight", col = "orange", border = "grey")

abline(v = active_ligand_target_links_df %>% 
         pull(weight) %>% 
         quantile(), 
       col = 'red', 
       lwd = 2, 
       lty = 2)
```


```{r}
active_ligand_target_links <- prepare_ligand_target_visualization(ligand_target_df = active_ligand_target_links_df, 
                                                                  ligand_target_matrix = ligand_target_matrix,
                                                                  cutoff = 0.25)

active_ligand_target_links %>% dim()
active_ligand_target_links[1:5, 1:5]
```

```{r}
order_ligands <- intersect(best_upstream_ligands, colnames(active_ligand_target_links)) %>% rev()
order_targets <- active_ligand_target_links_df$target %>% unique()

vis_ligand_target <- active_ligand_target_links[order_targets, order_ligands] %>% t()

p_ligand_target_network <- vis_ligand_target %>% 
  make_heatmap_ggplot(y_name = "Prioritized CAF-ligands", 
                      x_name = 'p-EMT genes in malignant cells', 
                      color = 'purple', 
                      legend_position = "top",
                      x_axis_position = 'top',
                      legend_title = "Regulatory potential") +
  scale_fill_gradient2(low = "whitesmoke", high = "purple", breaks = c(0, 0.005, 0.01)) +
  theme(axis.text.x = element_text(face = "italic"))

p_ligand_target_network
```


Note that the choice of these cutoffs for visualization is quite arbitrary. We recommend users to test several cutoff values.  

If you would consider more than the top 250 targets based on prior information, you will infer more, but less confident, ligand-target links; by considering less than 250 targets, you will be more stringent.  


## Follow-up analysis 1: Ligand-receptor network inference for top-ranked ligands  
One type of follow-up analysis is looking at which receptors of the receiver cell population (here: malignant cells) can potentially bind to the prioritized ligands from the sender cell population (here: CAFs).  

So, we will now infer the predicted ligand-receptor interactions of the top-ranked ligands and visualize these in a heatmap.  
```{r}
# get the ligand-receptor network of the top-ranked ligands  
lr_network_top <- lr_network %>% 
  filter(from %in% best_upstream_ligands & to %in% expressed_receptors) %>% 
  distinct(from, to)

best_upstream_receptors <- lr_network_top %>% 
  pull(to) %>% 
  unique()

# get the weights of the ligand-receptor interactions as used in th e NicheNet model
weighted_networks <- readRDS(file = "../../../Data/NicheNet/data/weighted_networks.rds")

lr_network_top_df <- weighted_networks$lr_sig %>% 
  filter(from %in% best_upstream_ligands & to %in% best_upstream_receptors)

# convert to matrix
lr_network_top_df <- lr_network_top_df %>% 
  spread(key = "from", value = "weight", fill = 0)

lr_network_top_matrix <- lr_network_top_df %>% 
  dplyr::select(-to) %>% 
  as.matrix() %>% 
  magrittr::set_rownames(lr_network_top_df$to)

# perform hierarchical clustering to order the ligands and receptors
dist_receptors <- dist(x = lr_network_top_matrix, method = "binary")
hclust_receptor <- hclust(d = dist_receptors, method = "ward.D2")
order_receptors <- hclust_receptor$labels[hclust_receptor$order]

dist_ligands <- dist(lr_network_top_matrix %>% t(), method = "binary")
hclust_ligands <- hclust(d = dist_ligands, method = "ward.D2")
order_ligands_receptor <- hclust_ligands$labels[hclust_ligands$order]
```


Show a heatmap of the ligand-receptor interactions  
```{r}
vis_ligand_receptor_network <- lr_network_top_matrix[order_receptors, order_ligands]
p_ligand_receptor_network <- vis_ligand_receptor_network %>% 
  t() %>% 
  make_heatmap_ggplot(y_name = "Prioritized CAF-ligands", 
                      x_name = "Receptors expresed by malignant cells", 
                      color = "mediumvioletred", 
                      x_axis_position = "top",
                      legend_title = "Prior interaction potential")

p_ligand_receptor_network
```


## Follow-up analysis 2: Visualize expression of top-predicted ligands and their target genes in a combined heatmap  
NicheNet only considers expressed ligands of sender cells, but does not take into account their expression for ranking the ligands. The ranking is purely based on the potential that a ligand might regulate the gene set of interest, given prior knowledge. Because it is also useful to further look into expression of ligands and their target genes, we demonstrate here how you could make a combined figure showing ligand activity, ligand expression, target gene expression and ligand-target regulatory potential.  

**Load additional packages required for the visualization**:  
```{r, message=FALSE}
library(RColorBrewer)
library(cowplot)
library(ggpubr)
```


**Prepare the ligand activity matrix**:  
```{r}
ligand_pearson_matrix <- ligand_activities %>% 
  select(pearson) %>% 
  as.matrix() %>% 
  magrittr::set_rownames(ligand_activities$test_ligand)

vis_ligand_pearson <- ligand_pearson_matrix[order_ligands, ] %>% 
  as.matrix(ncol = 1) %>% 
  magrittr::set_colnames("Pearson")
```

```{r}
p_ligand_pearson <- vis_ligand_pearson %>% 
  make_heatmap_ggplot(y_name = "Prioritized CAF-ligands", 
                      x_name = "Ligand activity", 
                      color = "darkorange", 
                      legend_position = "top",
                      legend_title = "Pearson correlation coefficient\ntarget gene prediction ability")
```

```{r}
vis_ligand_pearson %>%
  as.data.frame() %>% 
  mutate(genes = rownames(vis_ligand_pearson)) %>% 
  top_n(n = 20, w = Pearson) %>% 
  ggplot(aes(x = reorder(genes, Pearson), y = Pearson)) +
  geom_col(aes(fill = Pearson), position = "fill") +
  scale_y_continuous(name = "", expand = c(0, 0), labels = NULL) +
  scale_x_discrete(name = "Prioritized CAF-ligands", expand = c(0, 0)) +
  scale_fill_gradient(name = "Pearson correlation \ncoefficient target gene\n prediction activity", 
                      low = "white", high = "darkorange") +
  coord_flip() +
  theme(axis.ticks.length = unit(x = 0, units = "cm"), 
        legend.position = "top", 
        panel.grid.major = element_line(size = 0), 
        panel.background = element_rect(fill = NA))
```


**Prepare expression of ligands in fibroblast per tumor**:  
```{r}
expression_df_CAF <- expression[CAF_ids, order_ligands] %>% 
  data.frame() %>%  
  rownames_to_column("cell") %>% 
  tbl_df() %>% 
  inner_join(sample_info %>% select(cell, tumor), by = "cell")
```

```{r}
aggregated_expression_CAF <- expression_df_CAF %>% 
  group_by(tumor) %>% 
  select(-cell) %>% 
  summarise_all(mean)
```

```{r}
aggregated_expression_df_CAF <- aggregated_expression_CAF %>% 
  select(-tumor) %>% 
  t() %>% 
  magrittr::set_colnames(aggregated_expression_CAF$tumor) %>% 
  data.frame() %>% 
  rownames_to_column("ligand") %>% 
  tbl_df()
```

```{r}
aggregated_expression_matrix_CAF <- aggregated_expression_df_CAF %>% 
  select(-ligand) %>% 
  as.matrix() %>% 
  magrittr::set_rownames(aggregated_expression_df_CAF$ligand)
```

```{r}
order_tumors = c("HN6","HN20","HN26","HN28","HN22","HN25","HN5","HN18","HN17","HN16") 
# this order was determined based on the paper from Puram et al. Tumors are ordered according to p-EMT score.
vis_ligand_tumor_expression = aggregated_expression_matrix_CAF[order_ligands,order_tumors]
```


```{r}
library(RColorBrewer)
color <- colorRampPalette(colors = rev(brewer.pal(n = 7, name = "RdYlBu")))(100)
p_ligand_tumor_expression <- vis_ligand_tumor_expression %>% 
  make_heatmap_ggplot(y_name = "Prioritized CAF-ligands", x_name = "tumor", 
                      color = color[100], 
                      legend_position = "top", 
                      legend_title = "Expression\n(averaged over\nsingle cells)") +
  theme(axis.text.y = element_text(face = "italic"))
```

**Prepare expression of target genes in malignant cells per tumor**  
```{r}
expression_df_target = expression[maligant_ids, geneset_oi] %>% 
  data.frame() %>% 
  rownames_to_column("cell") %>% 
  tbl_df() %>% 
  inner_join(sample_info %>% 
  select(cell,tumor), by =  "cell") 

aggregated_expression_target = expression_df_target %>% 
  group_by(tumor) %>% 
  select(-cell) %>% 
  summarise_all(mean)

aggregated_expression_df_target = aggregated_expression_target %>% 
  select(-tumor) %>% 
  t() %>% 
  magrittr::set_colnames(aggregated_expression_target$tumor) %>% 
  data.frame() %>% 
  rownames_to_column("target") %>% 
  tbl_df() 

aggregated_expression_matrix_target = aggregated_expression_df_target %>%
  select(-target) %>% 
  as.matrix() %>% 
  magrittr::set_rownames(aggregated_expression_df_target$target)

vis_target_tumor_expression_scaled = aggregated_expression_matrix_target %>% 
  t() %>% 
  scale_quantile() %>% 
  .[order_tumors,order_targets]
```

```{r}
p_target_tumor_scaled_expression = vis_target_tumor_expression_scaled %>% 
  make_threecolor_heatmap_ggplot(y_name = "Tumor",
                                 x_name = "Target", 
                                 low_color = color[1], 
                                 mid_color = color[50], 
                                 mid = 0.5, 
                                 high_color = color[100], 
                                 legend_position = "top", 
                                 x_axis_position = "top" , 
                                 legend_title = "Scaled expression\n(averaged over\nsingle cells)") + 
  theme(axis.text.x = element_text(face = "italic"))

p_target_tumor_scaled_expression
```

**Combine the different heatmaps in one overvire figure**  
```{r}
figures_without_legend = plot_grid(
  p_ligand_pearson + theme(legend.position = "none", axis.ticks = element_blank()) + theme(axis.title.x = element_text()),
  p_ligand_tumor_expression + theme(legend.position = "none", axis.ticks = element_blank()) + theme(axis.title.x = element_text()) + ylab(""),
  p_ligand_target_network + theme(legend.position = "none", axis.ticks = element_blank()) + ylab(""), 
  NULL,
  NULL,
  p_target_tumor_scaled_expression + theme(legend.position = "none", axis.ticks = element_blank()) + xlab(""), 
  align = "hv",
  nrow = 2,
  rel_widths = c(ncol(vis_ligand_pearson)+ 4.5, ncol(vis_ligand_tumor_expression), ncol(vis_ligand_target)) -2,
  rel_heights = c(nrow(vis_ligand_pearson), nrow(vis_target_tumor_expression_scaled) + 3)) 

legends = plot_grid(
  as_ggplot(get_legend(p_ligand_pearson)),
  as_ggplot(get_legend(p_ligand_tumor_expression)),
  as_ggplot(get_legend(p_ligand_target_network)),
  as_ggplot(get_legend(p_target_tumor_scaled_expression)),
  nrow = 2,
  align = "h")

plot_grid(figures_without_legend, 
          #legends, 
          rel_heights = c(10, 2), 
          rel_widths = c(5, 8),
          nrow = 2,
          align = "hv")
```



